---
title: 'Syngnathus scovelli: Figures'
output:
  pdf_document: default
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,out.extra='',fig.pos='H', eval = FALSE, message=FALSE)
knitr::opts_knit$set(root.dir='./')
```

```{r loadLibs, message=FALSE, warning=FALSE,eval=TRUE}
library(spfTools)
library(DESeq2)
library(ggplot2)
library(pheatmap)
library(magick)
library(patchwork)
library(tidyverse)
library(UpSetR)
library(EnvStats)
library(knitr)
```

```{r setParams, eval=TRUE}
sex_bias_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray")

biased_bins <- c("Low", "Med", "High", "Extreme", "Sex.specific")
labs <- c("Low", "Med", "High", "Extreme", "Specific")

sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4" )
organ_shapes <- c(
  Brain=16, 
  Liver=17, 
  Ovaries=15,
  Testis=15,
  Gonad=15,
  Skin=19,
  BP=18)
sex_shapes <- c(F = 16, 
                M = 17)

organ_cols<-c("Brain" = "#6E8B3D", 
              "Gonad" = "#EEB422", 
              "Liver" = "#EE8262", 
              "Skin" = "#4F94CD", 
              "BP" = "#6959CD")
```

```{r getSamplesInfo, eval=TRUE}

#The samples file generated for tximport
samples <- read.table("plot_data/SS_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$Sex <- as.factor(samples$Sex)
samples$Organ <- as.factor(samples$Organ)

```


## PCA Plots Figure

For the PCA plots we have kept in the brood pouch samples, however, those are not included in the differential expression analysis.

```{r figPCADataPrep, eval=FALSE}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.salmon <- readRDS("../data/txi.salmon_SS.RDS")

#Create the DESeq dataset 
dds_PCA <- DESeqDataSetFromTximport(txi.salmon, 
                                   colData = samples,
                                   design = ~ Sex)
##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(dds_PCA)) >= 10 
dds_PCA <- dds_PCA[keep, ]

#Run the differential expression analysis
dds_PCA_exp <- DESeq(dds_PCA)


#Transform the data
vsd_PCA <- vst(dds_PCA_exp, blind=FALSE)
vsd_assay <- assay(vsd_PCA)
write.csv(vsd_assay,"plot_data/vsd_assay.csv", row.names=TRUE)

# get normalised counts
norm_counts<-counts(ddsMF_FL_exp, normalized = TRUE)
write.csv(norm_counts, "plot_data/normalized_counts.csv",row.names = TRUE)
```

```{r readCountData}
vsd_assay <- read.csv("plot_data/vsd_assay.csv",
                      row.names = 1)
```

```{r runPCA, eval=FALSE}
p <- pca(assay(vsd_PCA), 
         metadata = colData(dds_PCA))

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd_PCA)))
pca_plotting_data <- as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$Organ <- SS_samples$Organ
pca_plotting_data$Sex <- SS_samples$Sex

#Calculate the percent variation attributed to each axis 
percents <- round(p$variance, digits = 2)

write.csv(pca$rotation[, 1:4],
          "plot_data/pca_rotation.csv", 
          row.names=TRUE)

write.csv(pca_plotting_data,
          "plot_data/pca_plotting_data.csv", 
          row.names = TRUE)

write.table(percents,
            "plot_data/PCA_percents.txt", 
            row.names = FALSE)

```

```{r getPCAdata}
pca_plotting_data <- read.csv("plot_data/pca_plotting_data.csv",
                              row.names = 1)
percents <- unlist(read.delim("plot_data/PCA_percents.txt"))
pca_rotation <- read.csv("plot_data/pca_rotation.csv",
                         row.names = 1)
```


### Fig. PCA Panels A, B, and C - PCA Plots

```{r figPCAa}

pdf("figs/Fig_PCA1.pdf",height = 6,width=6)
par(mar=c(4,5,4,1), oma=c(2,2,2,2))
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC2,
     col = paste0(organ_cols[pca_plotting_data$Organ],"75"),
     pch = sex_shapes[pca_plotting_data$Sex],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percents[1], "% variance"),
     ylab = paste0("PC2: ",percents[2], "% variance"),
     bty = 'l',
     xpd = TRUE)
outer_legend("top",
             c("Female","Male"),
             pch = sex_shapes,
             bty = 'n',
             #col=paste0(sex_cols,"75"),
             cex = 2,
             ncol = 2,
             pt.cex = 2)

dev.off()
```

```{r figPCAb}

pdf("figs/Fig_PCA2.pdf", height = 6, width=6)
par(mar=c(4,5,4,1), oma=c(2,2,2,2))
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC3,
     col = paste0(organ_cols[pca_plotting_data$Organ],"75"),
     pch = sex_shapes[pca_plotting_data$Sex],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percents[1], "% variance"),
     ylab = paste0("PC3: ",percents[3], "% variance"),
     bty = 'l')

dev.off()
```

```{r figPCAc}

pdf("figs/Fig_PCA3.pdf", height = 6, width=6)
par(mar=c(4,5,4,1), oma=c(2,2,2,2))
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC4,
     col = paste0(organ_cols[pca_plotting_data$Organ],"75"),
     pch = sex_shapes[pca_plotting_data$Sex],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percents[1], "% variance"),
     ylab = paste0("PC4: ",percents[4], "% variance"),
     bty = 'l')

outer_legend("top",
             c("Brain","Gonad",
               "Liver", "Skin", 
               "Brood Pouch"),
             bty='n',
             pch = 16,
             col=paste0(organ_cols, "75"),
             cex=1.5,
             ncol=3,
             pt.cex=2)

dev.off()
```

### Fig. PCA Panels D, E, F and G - Heatmaps

```{r heatmapData}

df <- as.data.frame(samples[,c("Sex", "Organ")])
rownames(df) <- samples$ID
df$Organ <- as.character(df$Organ)

```


```{r heatmapSetup}
ann_colors = list(
  Sex=sex_cols,
  Organ=organ_cols
)

col_order<-c(
  rownames(df[df$Sex=="F"&df$Organ=="Brain",]),
  rownames(df[df$Sex=="M"&df$Organ=="Brain",]),
  rownames(df[df$Sex=="F"&df$Organ=="Gonad",]),
  rownames(df[df$Sex=="M"&df$Organ=="Gonad",]),
  rownames(df[df$Sex=="F"&df$Organ=="Liver",]),
  rownames(df[df$Sex=="M"&df$Organ=="Liver",]),
  rownames(df[df$Sex=="F"&df$Organ=="Skin",]),
  rownames(df[df$Sex=="M"&df$Organ=="Skin",]),
  rownames(df[df$Sex=="M"&df$Organ=="BP",])
)
  
```

```{r savePheatmapFxn}

# from https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

```

```{r figPCAd}

pc1 <- pheatmap(vsd_assay[which(abs(pca_rotation[,1]) >= 0.02),col_order], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df,
              annotation_colors=ann_colors,
              cellwidth = 9,
              fontsize = 16,
              annotation_legend = FALSE,
              main = "Top loading genes on PC1")

save_pheatmap_pdf(pc1, "figs/FigPCA_pc1_heatmap.pdf",width=6,height=6)

```


```{r figPCAe}

pc2 <- pheatmap(vsd_assay[which(abs(pca_rotation[,2]) >= 0.02),col_order], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df,
              annotation_colors=ann_colors,
              cellwidth = 9,
              fontsize = 16,
              annotation_legend = FALSE,
              main = "Top loading genes on PC2")

save_pheatmap_pdf(pc2, "figs/FigPCA_pc2_heatmap.pdf",width=6,height=6)

```


```{r figPCAf}
pc3 <- pheatmap(vsd_assay[which(abs(pca_rotation[,3]) >= 0.02),col_order], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df,
              annotation_colors=ann_colors,
              cellwidth = 9,
              fontsize = 16,
              annotation_legend = FALSE,
              border_color = NA,
              main = "Top loading genes on PC3")

save_pheatmap_pdf(pc3, "figs/FigPCA_pc3_heatmap.pdf",width=6,height=6)

```

```{r figPCAg}

pc4 <- pheatmap(vsd_assay[which(abs(pca_rotation[,4]) >= 0.02),col_order], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df,
              annotation_colors=ann_colors,
              cellwidth = 9,
              fontsize = 16,
              border_color = NA,
              main = "Top loading genes on PC4")

save_pheatmap_pdf(pc4, "figs/FigPCA_pc4_heatmap.pdf",width=7.2,height=6)

```

### Fig. PCA Assembly

```{r assembleFig2, message=FALSE}
figPCAa <- image_ggplot(image_read_pdf('figs/Fig_PCA1.pdf'),
                        interpolate = TRUE)
figPCAb <- image_ggplot(image_read_pdf('figs/Fig_PCA2.pdf'),
                        interpolate = TRUE)
figPCAc <- image_ggplot(image_read_pdf('figs/Fig_PCA3.pdf'),
                        interpolate = TRUE)
figPCAd <- image_ggplot(image_read_pdf('figs/FigPCA_pc1_heatmap.pdf'),
                        interpolate = TRUE)
figPCAe <- image_ggplot(image_read_pdf('figs/FigPCA_pc2_heatmap.pdf'),
                        interpolate = TRUE)
figPCAf <- image_ggplot(image_read_pdf('figs/FigPCA_pc3_heatmap.pdf'),
                        interpolate = TRUE)
figPCAg <- image_ggplot(image_read_pdf('figs/FigPCA_pc4_heatmap.pdf'),
                        interpolate = TRUE)

figPCA <- wrap_plots(
  figPCAa,
  figPCAb,
  figPCAc,
  figPCAd,
  figPCAe,
  figPCAf,
  figPCAg,
  design="ABC#
  DEFG"
)
figPCA <- figPCA + plot_annotation(tag_levels = 'A')

ggsave("figs/FigPCA.pdf",figPCA, height=4, width=5)
ggsave("figs/FigPCA.png",figPCA, height=4, width=5) # also save as a png
```



## Figure Sex-bias

```{r figSBData}

logFC_long <- read.csv("plot_data/logFC_long_sexbias.csv")
organs <- levels(as.factor(logFC_long$tissue))
ymax <- max(abs(logFC_long$logFC),
            na.rm = TRUE)+5

```

### Fig. Sex-bias boxplots

```{r figSB-boxplots}
pdf("figs/FigSB_logFC_boxplots.pdf", width = 10, height=4)
par(mfrow=c(1,4), 
    oma=c(4,5,2,8), 
    mar=c(1,1,1,0))

for(organ in organs){
  
  # add jittered points
  plot(abs(logFC_long$logFC[logFC_long$tissue==organ]) ~ 
           jitter(as.numeric(as.factor(logFC_long$bias[logFC_long$tissue==organ]))),
         col="#00000075", 
       axes=FALSE,
       cex.main=2,
       xlim=c(0,4),
       ylim=c(0,ymax)
       )
  
  # make the boxplot
  boxplot(
    abs(logFC_long$logFC[logFC_long$tissue==organ]) ~ logFC_long$bias[logFC_long$tissue==organ],
    col=scales::alpha(sex_bias_colors,0.75),
    add = TRUE,
    yaxt='n',
    las=1,
    cex.axis=1.75,
    frame=FALSE,
    lwd=1.5,
    outline=FALSE# do not plot outliers
  )
  # make the axis lines longer and thicker
  axis(1, labels=NA,at=-1:4,lwd=2, lwd.ticks=0, las = 3, cex.axis = 1.5)
  axis(2,labels=seq(-5,ymax,10),
       line=NA, at=seq(-5,ymax,10), 
       lwd=2,ylim=c(0,ymax),cex.axis=2,las=2)
  
  mtext(organ, cex=1.5,outer=FALSE, line=-1)
  
}

# add x axis label
mtext("Bias level",side = 1,cex=1.5, outer=TRUE, line=2)

# add y axis label
mtext(expression("|log"[2]*"FC|"),side = 2,cex=1.5, outer=TRUE, line=2.25)

# add legend
spfTools::outer_legend("right",
                       legend=c("Female\nbiased",
                                "Male\nbiased",
                                "Unbiased"),
                       pt.bg=sex_bias_colors,
                       pch=22,
                       bty='n',
                       ncol=1,
                       cex=2,
                       y.intersp = 1.5,
                       pt.cex=2)

dev.off()
```

### Fig. SB - categories

```{r figSB-cat-Data}
bias_cat_brain <- as.matrix(read.table("plot_data/bias_cat_brain.txt", 
                            header = TRUE, row.names = 1))
bias_cat_gonad <- as.matrix(read.table("plot_data/bias_cat_gonad.txt", 
                            header = TRUE, row.names = 1))
bias_cat_liver <- as.matrix(read.table("plot_data/bias_cat_liver.txt", 
                            header = TRUE, row.names = 1))
bias_cat_skin <- as.matrix(read.table("plot_data/bias_cat_skin.txt", 
                            header = TRUE, row.names = 1))

```

```{r figSB-cat}
pdf("figs/FigSB_biasCat_counts.pdf",width = 10, height=4)
ymax <- max(c(unlist(bias_cat_brain),
            unlist(bias_cat_gonad),
            unlist(bias_cat_skin),
            unlist(bias_cat_liver)))+500

par(mfrow=c(1, 4), oma=c(6,4,2,8), mar=c(1,2.5,1,0), 
    cex.main=2,
    cex.axis=2)

#Brain
bp<-barplot(bias_cat_brain[,biased_bins], 
        beside = TRUE,
        xaxt='n',
        ylim = c(0, max(bias_cat_brain)+10), 
        col = sex_cols,
        main = "")
mtext("Brain",3,outer = FALSE,cex=1.5,line=-1)
axis(2,lwd=2)
text(cex=2, x=colMeans(bp), y=-0.5, labs, xpd=NA, srt=35, adj = 1)

#Gonads
bp<-barplot(bias_cat_gonad[,biased_bins], 
        beside = TRUE, 
        ylim = c(0, ymax), 
        xaxt='n',
        col = sex_cols,
        main = "",
        cex.main=2,
        cex.axis=2)
mtext("Gonad",3,outer = FALSE,cex=1.5,line=-1)
axis(2,lwd=2,labels = NA)
text(cex=2, x=colMeans(bp), y=-145, labs, xpd=NA, srt=35, adj=1)

#Liver
barplot(bias_cat_liver[,biased_bins], 
        beside = TRUE, 
        ylim = c(0, max(bias_cat_liver)+50), 
        xaxt='n',
        col = sex_cols,
        main = "",
        cex.main=2,
        cex.axis=2)
axis(2,lwd=2, labels = NA)
text(cex=2, x=colMeans(bp), y=-2, labs, xpd=NA, srt=35,adj=1)
mtext("Liver",3,outer = FALSE,cex=1.5,line=-1)

#Skin
barplot(bias_cat_skin[,biased_bins], 
        beside = TRUE, 
        ylim = c(0, max(bias_cat_skin)+10), 
        xaxt='n',
        col = sex_cols,
        main = "",
        cex.main=2,
        cex.axis=2)
axis(2,lwd=2, labels = NA)
text(cex=2, x=colMeans(bp), y=-5, labs, xpd=NA, srt=35,adj=1)
mtext("Skin",3,outer = FALSE,cex=1.5,line=-1)

mtext("Number of Genes",2,outer=TRUE, cex=1.5, line=2.25)
mtext("Bias category",1, outer=TRUE, cex=1.5, line=4)

outer_legend("right", 
       legend = c("Female\nbiased", "Male\nbiased"), 
       pt.bg = sex_cols,
       pch=22,
       bty='n',
       ncol=1,
       cex=2,
       y.intersp = 1.5,
       pt.cex=2)
dev.off()
```

### Fig. SB - Upset plot

```{r fig3cData}
gonad_mal_biased <- read.table("plot_data/gonad_mal_biased.txt", row.names = 1)
gonad_fem_biased <- read.table("plot_data/gonad_fem_biased.txt", row.names = 1)
brain_mal_biased <- read.table("plot_data/brain_mal_biased.txt", row.names = 1)
brain_fem_biased <- read.table("plot_data/brain_fem_biased.txt", row.names = 1)
liver_mal_biased <- read.table("plot_data/liver_mal_biased.txt", row.names = 1)
liver_fem_biased <- read.table("plot_data/liver_fem_biased.txt", row.names = 1)
skin_mal_biased <- read.table("plot_data/skin_mal_biased.txt", row.names = 1)
skin_fem_biased <- read.table("plot_data/skin_fem_biased.txt", row.names = 1)

listInputall <- list("MB Gonad" = rownames(gonad_mal_biased),
                     "MB Brain"=rownames(brain_mal_biased), 
                     "MB Liver"=rownames(liver_mal_biased),
                     "MB Skin"=rownames(skin_mal_biased),
                     "FB Gonad" = rownames(gonad_fem_biased),
                     "FB Brain"=rownames(brain_fem_biased), 
                     "FB Liver"=rownames(liver_fem_biased),
                     "FB Skin"=rownames(skin_fem_biased))

```

```{r fig3c, eval=FALSE}
# color added via illustrator
pdf("figs/FigSB_sexbias_upset.pdf",width = 10, height=7)
upset(fromList(listInputall),
      mainbar.y.label = "# Shared Sex-Biased Genes",
      sets.x.label = "# Sex-Biased Genes",
      point.size = 3,
      nsets = 8,
      nintersects = NA,
      text.scale = c(2, 2, 2, 1.5, 2, 1.5)
)
dev.off()

```

### Fig. SB Assembly

```{r fig3, message=FALSE}

figSBa <- image_ggplot(image_read_pdf('figs/FigSB_logFC_boxplots.pdf'),interpolate = TRUE)
figSBb <- image_ggplot(image_read_pdf('figs/FigSB_biasCat_counts.pdf'),interpolate = TRUE)
figSBc <- image_ggplot(image_read_pdf('figs/FigSB_sexbias_upset_color.pdf'),interpolate = TRUE)


figSB <- wrap_plots(
  figSBa,
  figSBb,
  figSBc,
  nrow = 3
  #design=design
)

figSB <- figSB + plot_annotation(tag_levels = 'A')
figSB

ggsave("figs/FigSB.pdf",figSB, height=10, width=6)
ggsave("figs/Fig3.png",fig3, height=6, width=8)

```

```{r showAssembledFig3, eval=TRUE}
knitr::include_graphics("figs/Fig3.png")
```

## Figure Tau

```{r figTAUData}

logFC_long_tauSB <- read.csv("plot_data/logFC_long_taubiasSB.csv")
logFC_long_tau <- read.csv("plot_data/logFC_long_taubias_SS.csv")
logFC_long_tau <- logFC_long_tau[logFC_long_tau$bias_cat != "Sex-specific",]
logFC_long_tau$tissue <- factor(logFC_long_tau$tissue, 
                                levels = c("Gonad", "Skin", 
                                           "Liver", "Brain"), 
                                ordered = TRUE)

organs <- levels(as.factor(logFC_long_tau$tissue))

organ_cols<-c("Brain" = "#6E8B3D", 
              "Gonad" = "#EEB422", 
              "Liver" = "#EE8262", 
              "Skin" = "#6959CD")

```

### Fig. Tau - tau v bias

```{r figtauvSB-SBG}
pdf("figs/Fig_tau_sexbias_SBG.pdf",width = 8, height=5)
par(oma=c(2,2,1,2),
    mar=c(2,2,1,2),
    xpd=FALSE)

plot(logFC_long_tauSB$tau[logFC_long_tauSB$tissue=="Gonad"]~
       sqrt(abs(logFC_long_tauSB$logFC[logFC_long_tauSB$tissue=="Gonad"])),
     xlim=c(min(sqrt(abs(logFC_long_tauSB$logFC))),5),
     ylim=c(0,1),
     xlab="",
     ylab="",
     bty="n",
     type='n',
     axes=FALSE
)

axis(1,pos=0,lwd=2,cex=2, cex.axis=1.5, las=1)
axis(2,pos=min(sqrt(abs(logFC_long_tauSB$logFC))),lwd=2,cex=2, cex.axis=1.5, las=1)
clip(min(sqrt(abs(logFC_long_tauSB$logFC))),8,0,1)

for(organ in organs){
  
  points(logFC_long_tauSB$tau[logFC_long_tauSB$tissue==organ]~
           sqrt(abs(logFC_long_tauSB$logFC[logFC_long_tauSB$tissue==organ])),
         col=paste0(organ_cols[organ],"50"),
         pch=19)

}

for(organ in organs){

  abline(lm(logFC_long_tauSB$tau[logFC_long_tauSB$tissue==organ]~
              sqrt(abs(logFC_long_tauSB$logFC[logFC_long_tauSB$tissue==organ]))),
         col=organ_cols[organ],
         lwd=3,
         lty=which(organs %in% organ),
         xpd=FALSE)
}

outer_legend("top",
       names(organ_cols[order(names(organ_cols))]),
       col=organ_cols[order(names(organ_cols))],
       pch=19,
       lwd=3,
       bty='n',
       cex=1.5,
       lty=1:4,
       ncol=4
)

mtext(expression("|log"[2]*"fold-change|"), 1, cex=1.75, line=2.5)
mtext(expression(tau["TPM"]),2,cex=2, line=2.5)
dev.off()
```

```{r figtauvSB-all}
pdf("figs/Fig_tau_sexbias_all.pdf",width = 8, height=5)
par(oma=c(2,2,1,2),
    mar=c(2,2,1,2),
    xpd=FALSE)

plot(logFC_long_tau$tau[logFC_long_tau$tissue=="Gonad"]~
       sqrt(abs(logFC_long_tau$logFC[logFC_long_tau$tissue=="Gonad"])),
     xlim=c(0,5),
     ylim=c(0,1),
     xlab="",
     ylab="",
     bty="n",
     type='n',
     axes=FALSE
)

axis(1,pos=0,lwd=2,cex=2, cex.axis=1.5, las=1)
axis(2,pos=0,lwd=2,cex=2, cex.axis=1.5, las=1)
clip(0,8,0,1)

for(organ in organs){
  
  points(logFC_long_tau$tau[logFC_long_tau$tissue==organ]~
           sqrt(abs(logFC_long_tau$logFC[logFC_long_tau$tissue==organ])),
         col=paste0(organ_cols[organ],"50"),
         pch=19)

}

for(organ in organs){

  abline(lm(logFC_long_tau$tau[logFC_long_tau$tissue==organ]~
              sqrt(abs(logFC_long_tau$logFC[logFC_long_tau$tissue==organ]))),
         col=organ_cols[organ],
         lwd=3,
         lty=which(organs %in% organ),
         xpd=FALSE)
}

outer_legend("top",
       names(organ_cols[order(names(organ_cols))]),
       col=organ_cols[order(names(organ_cols))],
       pch=19,
       lwd=3,
       bty='n',
       cex=1.5,
       lty=1:4,
       ncol=4
)

mtext(expression("|log"[2]*"fold-change|"), 1, cex=2, line=2.5)
mtext(expression(tau["TPM"]),2,cex=2, line=2.5)
dev.off()
```

### Fig. Tau - tau v. bias categories

```{r figtau-catsetup}
bias_labs <- c("U","L", "M", "H", "E", "S")
bias_bins <- c("Unbiased",biased_bins)
```

```{r figtau-catData}
logFC_long_all_ss <- read.csv("plot_data/logFC_long_taubias_SS.csv")

logFC_long_all_ss$bias_cat <- factor(logFC_long_all_ss$bias_cat,
                                  levels = bias_bins, ordered = TRUE)

```


```{r figtau-cat}

pdf("figs/Fig_tau_biascat_violins.pdf",width = 15, height=3.75)

logFC_long_all_ss %>%
  ggplot(aes(x = bias_cat, y = tau, fill = bias)) +
  geom_violin(position = position_dodge(), draw_quantiles = c(0.5)) +
  scale_x_discrete(labels= bias_labs) +
  geom_boxplot(width = 0.1, color = "black", position = position_dodge(width = 0.9)) +
  scale_fill_manual(values = sex_bias_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        strip.background = element_blank(),
        axis.line = element_line(colour = "black"),
        text=element_text(size=16)) +
  
  labs(x="Bias Category", y=expression(tau["TPM"]))  +
  stat_n_text(data = logFC_long_all_ss[logFC_long_all_ss$bias == "FB",], 
              aes(x=bias_cat, y=tau),
              #hjust = 1.2,
              #vjust = -2,
              y.pos = -0.05,
              color=sex_bias_colors["FB"], size = 3
              ) +
  stat_n_text(data = logFC_long_all_ss[logFC_long_all_ss$bias == "MB",], 
              aes(x=bias_cat, y=tau),
             # hjust = -0.2,
              #vjust = 2,
              y.pos = 0.95,
             color=sex_bias_colors["MB"], size = 3
              ) +
  stat_n_text(data = logFC_long_all_ss[logFC_long_all_ss$bias == "UB",], 
              aes(x=bias_cat, y=tau),
              #vjust = -4 ,
              y.pos=-0.05,
              color=sex_bias_colors["UB"], size = 3
              ) +
  guides(fill = guide_legend(title = "Bias", order = 3))

dev.off()
```

### Fig. Tau Assembly

```{r fig4, message=FALSE}
figTAUa <- image_ggplot(image_read_pdf('figs/Fig_tau_sexbias_SBG.pdf'),interpolate = TRUE)
figTAUb <- image_ggplot(image_read_pdf('figs/Fig_tau_sexbias_all.pdf'),interpolate = TRUE)
figTAUc <- image_ggplot(image_read_pdf('figs/Fig_tau_biascat_violins.pdf'),interpolate = TRUE)

figTAU <- plot_grid(plot_grid(figTAUa, figTAUb, ncol = 2, labels = "AUTO"),
          figTAUc,
          nrow = 2,
          labels = c("", "C"))

ggsave("figs/FigTAU.pdf",figTAU, height = 6, width=11)
ggsave("figs/FigTAU.png",figTAU, height = 6, width=11)

```

```{r showAssembledFig4, eval=TRUE}
knitr::include_graphics("figs/FigTAU.png")
```

## Figure Gene Ontology - All Organs

```{r figGOdata}
all_go_sums_SBG <- read.csv("plot_data/GO_freq_SBG.csv")
all_go_sums_SS <- read.csv("plot_data/GO_freq_SS.csv")
```

### Fig. GO - all sex-biased

```{r figGO-sbg}
pdf("figs/FigGO_SBGbarplot.pdf",height=16, width=12)

all_go_sums_SBG$prot_class <- str_wrap(all_go_sums_SBG$prot_class, 
                                       width = 31)
all_go_sums_SBG$prot_class[all_go_sums_SBG$prot_class=="unclassified"]<- " unclassified"
all_go_sums_SBG$prot_class[all_go_sums_SBG$prot_class=="other"]<- " other"

ggplot(all_go_sums_SBG, 
       aes(fct_rev(prot_class), prop, fill = tissue)) +   
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = organ_cols) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=20),
        axis.text.y = element_text(size=15, hjust = 0.5, 
                                   margin = margin(b = 20),
                                   vjust = 0.75),
        text=element_text(size=20),
        legend.position = "bottom",
        axis.ticks.length.y = unit(.25, "cm")) +
  coord_flip() + 
  labs(y="Proportion of sex-biased genes", x="")

dev.off()

```

### Fig. GO - all sex-specific

```{r figGO-ss}
pdf("figs/FigGO_SSbarplot.pdf",height=16, width=12)

all_go_sums_SS$prot_class <- str_wrap(all_go_sums_SS$prot_class, 
                                       width = 31)
all_go_sums_SS$prot_class[all_go_sums_SS$prot_class=="unclassified"]<- " unclassified"
all_go_sums_SS$prot_class[all_go_sums_SS$prot_class=="other"]<- " other"

ggplot(all_go_sums_SS, 
       aes(fct_rev(prot_class), prop, fill = tissue)) +   
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = organ_cols) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=20),
        axis.text.y = element_text(size=20, hjust = 0.5, 
                                   margin = margin(b = 20),
                                   vjust = 0.75),
        text=element_text(size=20),
        legend.position = "bottom",
        axis.ticks.length.y = unit(.25, "cm")) +
  coord_flip() + 
  labs(y="Proportion of sex-specific genes", x="")

dev.off()

```

### Fig. GO Assembly

```{r figGO, message=FALSE}

figGOa <- image_ggplot(image_read_pdf('figs/FigGO_SBGbarplot.pdf'),
                       interpolate = TRUE)
figGOb <- image_ggplot(image_read_pdf('figs/FigGO_SSbarplot.pdf'),
                       interpolate = TRUE)

figGO <- wrap_plots(figGOa,
                    figGOb,
                    ncol = 2)

figGO <- figGO + plot_annotation(tag_levels = 'A')

figGO

ggsave("figs/FigGO.pdf", figGO, height=6, width=8)
ggsave("figs/FigGO.png", figGO, height=6, width=8)

```

## Figure Gene Ontology - Just the Gonads

```{r figGOdata-gonads}
gon_go_sums_SBG <- read.csv("plot_data/GO_freq_SBG_Gonads.csv")
```

```{r figGO-sbg-gonads}
pdf("figs/FigGO_SBGbarplot_GON.pdf",height=10, width=16)

gon_go_sums_SBG$prot_class[gon_go_sums_SBG$prot_class=="unclassified"]<- " unclassified"
gon_go_sums_SBG$prot_class[gon_go_sums_SBG$prot_class=="other"]<- " other"

ggplot(gon_go_sums_SBG, 
       aes(fct_rev(prot_class), prop, fill = tissue)) +   
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = organ_cols) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=20),
        axis.text.y = element_text(size=15, hjust = 0.5, 
                                   margin = margin(b = 20),
                                   vjust = 0.75),
        text=element_text(size=20),
        legend.position = "bottom",
        axis.ticks.length.y = unit(.25, "cm")) +
  coord_flip() + 
  labs(y="Proportion of sex-biased genes", x="")

dev.off()

```

## Figure Tajima's D vs Sex bias Category

```{r figTajDdata}
#Read in the data that contains info about Tajima's D and Sex bias
logFC_TD <- read.csv("plot_data/sexbias_tajD.csv")

#Make sure the Bias cats. plot in the correct order
logFC_TD$bias <- factor(logFC_TD$bias, 
                        levels = c("FB", "NB", "MB"), 
                        ordered = TRUE)

#Set the max and min y-axis value based on the TD so they are all on the same scale
ymax <- max(logFC_TD$Tajima)+0.5
ymin <- min(logFC_TD$Tajima)-0.5

#Create a vector that contains the organs we want to investigate (No Brain)
organs <- c("Gonad", "Liver", "Skin")

#Set colors
sex_bias_colors <- c("FB" = "#7fc97f", "NB" = "darkgray", "MB" = "#beaed4")

```


### Fig. Tajima's D boxplots

```{r figTD-boxplots}
pdf("figs/FigTD_bias_boxplots.pdf", width = 10, height=4)
par(mfrow=c(1,3), 
    oma=c(4,4,2,8), 
    mar=c(1,1,1,0))

for(organ in organs){
  
  # add jittered points
  plot(logFC_TD$Tajima[logFC_TD$tissue==organ] ~ 
           jitter(as.numeric(as.factor(logFC_TD$bias[logFC_TD$tissue==organ]))),
         col="#00000075", 
       axes=FALSE,
       cex.main=2,
       xlim=c(0,4),
       ylim=c(ymin, ymax)
       )
  
  # make the boxplot
  boxplot(logFC_TD$Tajima[logFC_TD$tissue==organ] ~ logFC_TD$bias[logFC_TD$tissue==organ],
    col=scales::alpha(sex_bias_colors,0.75),
    add = TRUE,
    yaxt='n',
    las=1,
    cex.axis=1.75,
    frame=FALSE,
    lwd=1.5,
    outline=FALSE# do not plot outliers
  )
  # make the axis lines longer and thicker
  axis(1, labels=NA,at=-1:4,lwd=2, lwd.ticks=0, las = 3, cex.axis = 1.5)
  axis(2, 
       labels = seq(-3, ymax, 1),
       line=NA, 
       at = seq(-3, ymax, 1), 
       lwd=2,
       ylim=c(ymin,ymax),
       cex.axis=2,las=2)
  
  mtext(organ, cex=1.5,outer=FALSE, line=-1)
  
}

# add x axis label
mtext("Bias level",side = 1,cex=1.5, outer=TRUE, line=2)

# add y axis label
mtext("Tajima's D",side = 2,cex=1.5, outer=TRUE, line=2.25)

# add legend
spfTools::outer_legend("right",
                       legend=c("Female\nbiased",
                                "Male\nbiased",
                                "Unbiased"),
                       pt.bg=sex_bias_colors[c(1,3,2)],
                       pch=22,
                       bty='n',
                       ncol=1,
                       cex=2,
                       y.intersp = 1.5,
                       pt.cex=2)

dev.off()
```

### Fig. Tajima's D - TD versus FC

```{r figtauvSB-SBG}
pdf("figs/Fig_TDvFC.pdf",width = 10, height=6)

par(oma=c(2,2,1,2),
    mar=c(2,2,1,2),
    xpd=FALSE)

plot(logFC_TD$Tajima[logFC_TD$tissue=="Gonad"]~
       logFC_TD$logFC[logFC_TD$tissue=="Gonad"],
     xlab="",
     ylab="",
     bty="n",
     type='n',
     axes=FALSE)

axis(1,pos=min(logFC_TD$Tajima)-0.1,lwd=2,cex=2, cex.axis=1.5, las=1)
axis(2,pos=min(logFC_TD$logFC)+2,lwd=2,cex=2, cex.axis=1.5, las=1)
clip(-10, 15.5, -3, 4)

points(logFC_TD$Tajima[logFC_TD$tissue=="Gonad"] ~ 
         logFC_TD$logFC[logFC_TD$tissue=="Gonad"],
       col=ifelse(logFC_TD$bias[logFC_TD$tissue=="Gonad"] == "FB",
                  "#7fc97f75",
                  ifelse(logFC_TD$bias[logFC_TD$tissue=="Gonad"] == "MB",
                         "#beaed475",
                         "#A9A9A975")),
       pch = 19)

model <- lm(Tajima ~ poly(logFC, 3), 
                   data = logFC_TD[logFC_TD$tissue == "Gonad", ])
lines(sort(logFC_TD$logFC[logFC_TD$tissue == "Gonad"]),
      fitted(model)[order(logFC_TD$logFC[logFC_TD$tissue == "Gonad"])],
      type = "l",
      lwd = 3,
      lty = 2)


outer_legend("top",
             legend = c("Female-biased", "Male-biased", "Unbiased"),
             pt.bg=sex_bias_colors[c(1,3,2)],
             pch = 22,
             ncol=3,
             bty = 'n',
             cex = 1.5,
             pt.cex = 2)

mtext(expression("log"[2]*"fold-change"), 1, cex=1.75, line=2.5)
mtext("Tajima's D", 2, cex=1.75, line=-0.5)

dev.off()
```

### Fig. Tajima's Assembly

```{r figTD, message=FALSE}
figTDa <- image_ggplot(image_read_pdf('figs/FigTD_bias_boxplots.pdf'),interpolate = TRUE)
figTDb <- image_ggplot(image_read_pdf('figs/Fig_TDvFC.pdf'),interpolate = TRUE)

figTD <- plot_grid(figTDa,
                   figTDb,
                   nrow = 
          figTAUc,
          nrow = 2,
          labels = c("", "C"))
figTD <- wrap_plots(figTDa,
                    figTDb,
                    nrow = 2)

figTD <- figTD + plot_annotation(tag_levels = 'A')
#figTD

ggsave("figs/FigTD.pdf",figTD, height = 8, width=9)
ggsave("figs/FigTD.png",figTD, height = 8, width=9)

```

```{r showAssembledFig4, eval=TRUE}
knitr::include_graphics("figs/FigTD.png")
```