---
title: "Differential Expression Analysis in _Syngnathus scovelli_"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
```

``` {r library, include = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(ggplot2)
library(ggpubr)
#library(cowplot)
library(UpSetR)
library(kableExtra)
library(tidyverse)
library(knitr)
```

```{r function}
# from https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
SS_txi.salmon <- readRDS("data/txi.salmon_SS.RDS")

#The samples file generated for tximport
SS_samples <- read.table("SS_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
SS_samples$Sex <- as.factor(SS_samples$Sex)
SS_samples$Organ <- as.factor(SS_samples$Organ)

#Add a column to sample data that contains the Submission number for each project
SS_samples$sub_num <- c(rep("PRJNA295858", 10),
                        rep("SRA582537", 7),
                        rep("SRA1439291", 5),
                        rep("SRA1651277", 4),
                        rep("SRA1439291", 5),
                        rep("SRA966117", 10))
```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females
To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions. I am starting with a simple single-factor analysis that incorporates all of the samples across organs to look at differences in expression between males and females. This analysis is run as counts ~ Sex.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_SS <- DESeqDataSetFromTximport(SS_txi.salmon, 
                                   colData = SS_samples,
                                   design = ~ Sex)

```

Prior to running the analysis I want to pre-filter the dataset to remove any rows with low counts. The cutoff I am using here is to remove rows that have a row sum of less than 10.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_SS)) >= 10
dds_SS <- dds_SS[keep, ]

```

After filtering I can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_SS_exp <- DESeq(dds_SS)

#Compile the results, using a p-value cut off of 0.05
res <- results(dds_SS_exp, alpha = 0.05)
summary(res)

```

From the summary of our results we can see that there are slightly more female-biased (LFC < 0) genes than there are male-biased genes (LFC > 0), but not by too much.

## Visualizing the results of the Single-Factor analysis across organs
### MA-plot: Looking at MvF differential expression
I am first going to generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.05.

```{r MA-plot, echo=FALSE, fig.cap="LogFC versus the mean normalized count for all of the genes. Points that are blue have a p-value less than 0.05. The two plots are the same with the only difference coming from an adjusted y-limit. Points that are triangles represent genes with a fold change higher/lower than the y-limit."}

par(mfrow=c(1,2))
plotMA(res, ylim = c(-2,2))
plotMA(res)

```

### Heatmap: Looking at patterns in overall expression
We can also generate a heat map to look at overall expression levels across our samples. Note, this is not differentially expressed genes.

```{r heatmap, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels."}
#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_SS_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds_SS)[,c("Sex", "Organ")])

#Transform the data
vsd <- vst(dds_SS_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df)
```

From the heatmap (Fig. \@ref(fig:heatmap)) we can see that for a lot of the top expressed genes the highest expression is foundin the liver, and for some of those, specifically in the female livers. There appears to be one gene that is highly expressed across all tissues and a few genes that are highly expressed in the skeletal/muscular tissues (brood pouches and skin). With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
#Run once to create the file and then commented out
#write.table(heatmap_TG,
 #           'data/heatmap_trinitygenes.txt',
 #           sep = "",
 #           quote=FALSE,
 #           row.names = FALSE,
 #           col.names = FALSE)
```

### Sample-dist Heatmap
We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples.

```{r sample-dist, echo=FALSE, fig.cap="Sample-to-sample distances heatmap showing the similarities and differences between all of the samples. The darker the color, the more similar the samples are. The diagonal line of very dark blue represents the comparissons between the same samples.", warning=FALSE}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")
colnames(sampleDistsMatrix) <- vsd$ID

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

We can see that the highest similarities (aside from same samples comparisons) is between samples from the same organ (Fig. \@ref(fig:sample-dist)). After that we can see high similarities between the brood pouch samples and the skin samples, and between the male gonads and the female gonads.


### PCA plots
Next we can look at PCA plots for our samples to see how our different samples are grouping.

```{r pca-pairs-plot, echo=FALSE, fig.cap="Principal components analysis pairsplot showing PCA axis 1 - 4."}
#Generate the PCA data to create the pairs plot, using the PCAtools package
p <- pca(assay(vsd), 
         metadata = colData(dds_SS))

#Create the pairs plot
pairsplot(p,
          components = getComponents(p, c(1:4)),
          colby = "Organ",
          triangle = FALSE,
          hline = 0, vline = 0,
          colkey = c("Brain" = "#6E8B3D75", 
                     "Gonad" = "#EEB42275", 
                     "Liver" = "#EE826275", 
                     "Skin" = "#4F94CD75",
                     "BP" = "#6959CD75"),
          shape = "Sex",
          shapekey = c(16, 17),
          pointSize = 3,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'),
          legendPosition = "left", legendIconSize = 3, legendLabSize = 10)

```

From the pairs plot we can see that PC1 appears to be explaining differences between the liver and the other tissues, PC2 appears to be explaining differences between muscular/skeletal tissues (skin and BP) and the other tissues, PC3 seems to account for differences between gonad and somatic tissues and PC4 is where we see some distinct separating of male and female samples, especially in the gonads. I am going to plot all of PC2 - 4 against PC1 below for a clearer picture.

```{r pca1v2, echo=FALSE, fig.cap="Principal components analysis reflects that most variation corresponds to differences in expression between organs (green: brain; yellow: gonad; pink: liver, blue: skin, purple: brood pouch), rather than variation due to sex (circle = female; triangle = male). The first axis explains 26% of variation in the dataset and the second axis explains 22% of the variation in gene expression with gonads, liver, and brain on the opposite side from the broodpouch and skin tissues.", warning=FALSE}

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$Organ <- SS_samples$Organ
pca_plotting_data$Sex <- SS_samples$Sex

#Calculate the percent variation attributed to each axis 
percentVar <- round(p$variance, digits = 2)


my_colors <- c("Brain" = "#6E8B3D75", "Gonad" = "#EEB42275", "Liver" = "#EE826275", "Skin" = "#4F94CD75", "BP" = "#6959CD75")
my_shapes <- c(16, 17)

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

```{r pca1v3, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC3. This third axis explains 16% of the variation, largely attributed to differences between the gonads and somatic tissues.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: ",percentVar[3],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

```{r pca1v4, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC4. This fourth axis explains 6.5% of the variation, largely attributed to differences between sexes in the gonads.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC4, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

# Investigating the Multi-factor analysis
Because these samples come from a variety of sources and have been sequenced across a variety of institutes at various times and with various parameters, including all of the samples in the DESeq2 analysis (and therefore normalizing across all of the samples) may be influencing outcomes. 

To test this I am going to set up the multi-factor analysis as I normally would and then I am going to perform a single-factor analysis within one organ to see if the outcomes are drastically different.

First, lets run the multi-factor analysis as counts ~ group, where group includes both the organ and sex (i.e. MLiver, Fliver, etc.). I am removing the brood pouch samples from this analysis as I cannot compare across the sexes (only males have a brood pouch).

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
#Create an additional column the has the two conditions combined(sex and organ type)
SS_samples$group <- factor(paste0(SS_samples$Sex, SS_samples$Organ))

##Create a copy of the DESeq dataset
ddsMF_SS <- DESeqDataSetFromTximport(SS_txi.salmon,
                                     colData = SS_samples,
                                     design = ~ group)

##Remove the brood pouch samples from the dataset
ddsMF_SS <- ddsMF_SS[, ddsMF_SS$Organ != "BP"]

ddsMF_SS$group <- as.character(ddsMF_SS$group)
ddsMF_SS$group <- as.factor(ddsMF_SS$group)

##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(ddsMF_SS)) >= 10 
ddsMF_SS <- ddsMF_SS[keep, ]

#Run the differential expression analysis
ddsMF_SS_exp <- DESeq(ddsMF_SS)
resultsNames(ddsMF_SS_exp)

```

## Look at the results of the MF analysis in the Liver
There are many criteria that have been employed previously as cut offs for what is sex biased vs what is not. For this study I am classifying sex-biased genes as: **genes with a p-value < 0.05 AND a logFC $\ge |2|$**.

```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
liver_con_res <- results(ddsMF_SS_exp, contrast = c("group", "MLiver", "FLiver"),
                         alpha = 0.05)
liver_con_res$trin_geneid <- rownames(liver_con_res)
head(liver_con_res)

summary(liver_con_res)

#Number of sex-biased genes based on our criteria
#Male-biased
sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange >= 2, 
    na.rm = TRUE)

#Female-biased
sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange <= -2, 
    na.rm = TRUE)
```

## Running a single factor analysis comparing M-F expression WITHIN the liver only
```{r liver-sf-analysis, message=FALSE}
#Create a subset of the DESeq dataset that only includes the liver samples
dds_SS_liver <- dds_SS[, dds_SS$Organ == "Liver"]

#Filtering to remove any rows that have less than 10 reads total
keep <- rowSums(counts(dds_SS_liver)) >= 10
dds_SS_liver <- dds_SS_liver[keep, ]

design(dds_SS_liver) <- ~ Sex

#Run the differential expression analysis
dds_SS_liver_exp <- DESeq(dds_SS_liver)

#Compliling the results
res <- results(dds_SS_liver_exp, alpha = 0.05)

summary(res)

#Number of sex-biased genes based on our criteria
#Male-biased
sum(res$padj <= 0.05 & res$log2FoldChange >= 2, 
    na.rm = TRUE)

#Female-biased
sum(res$padj <= 0.05 & res$log2FoldChange <= -2, 
    na.rm = TRUE)
```

There doesn't appear to be a huge difference between the number of sex-biased genes that are highlighted when the liver is investigated within the multi-factor analysis or when it is investigated as it's own single factor analysis. For a more conservative estimate of sex-bias within each organ I am going to move forward with the single factor analysis within each organ type. 

# Running a single factor analysis between males and females WITHIN each organ
## Liver M-F differential expression
The single factor analysis from the liver showed `r sum(res$padj <= 0.05 & res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(res$padj <= 0.05 & res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. I am going to pull out all of the female- and male-biased genes to store them in their own objects as well as genes that are non-biased. Non-biased genes will be categorized as genes with a p-value > 0.05, regardless of what the logFC may be.

```{r liver-sf-sexbiased}
#Removing the rows where padj. is NA in results
liver_con_res_noNA <- res[!is.na(res$padj),]
summary(liver_con_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed
#write.csv(as.data.frame(liver_con_res_noNA), "data/liver_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
liver_mal_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange >= 2
                                             & liver_con_res_noNA$padj <= 0.05),]
liver_fem_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange <= -2 
                                             & liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
liver_non_biased <- liver_con_res_noNA[which(liver_con_res_noNA$padj > 0.05),]
```

## Brain M-F differential expression

```{r brain-sf-analysis, message=FALSE}
#Create a subset of the DESeq dataset that only includes the liver samples
dds_SS_brain <- dds_SS[, dds_SS$Organ == "Brain"]

#Filtering to remove any rows that have less than 10 reads total
keep <- rowSums(counts(dds_SS_brain)) >= 10
dds_SS_brain <- dds_SS_brain[keep, ]

design(dds_SS_brain) <- ~ Sex

#Run the differential expression analysis
dds_SS_brain_exp <- DESeq(dds_SS_brain)

#Compiling the results
res_brain <- results(dds_SS_brain_exp, alpha = 0.05)

summary(res_brain)
```

In the brain there are `r sum(res_brain$padj <= 0.05 & res_brain$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(res_brain$padj <= 0.05 & res_brain$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. Let's now save those and the non biased genes into their own objects.

```{r brain-sf-sexbiased}
#Removing the rows where padj. is NA in results
brain_res_noNA <- res_brain[!is.na(res_brain$padj),]
summary(brain_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed
#write.csv(as.data.frame(brain_res_noNA), "data/brain_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the brain
brain_mal_biased <- brain_res_noNA[which(brain_res_noNA$log2FoldChange >= 2
                                             & brain_res_noNA$padj <= 0.05),]
brain_fem_biased <- brain_res_noNA[which(brain_res_noNA$log2FoldChange <= -2 
                                             & brain_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the brain
brain_non_biased <- brain_res_noNA[which(brain_res_noNA$padj > 0.05),]
```

## Skin M-F differential expression

```{r skin-sf-analysis, message=FALSE}
#Create a subset of the DESeq dataset that only includes the liver samples
dds_SS_skin <- dds_SS[, dds_SS$Organ == "Skin"]

#Filtering to remove any rows that have less than 10 reads total
keep <- rowSums(counts(dds_SS_skin)) >= 10
dds_SS_skin <- dds_SS_skin[keep, ]

design(dds_SS_skin) <- ~ Sex

#Run the differential expression analysis
dds_SS_skin_exp <- DESeq(dds_SS_skin)

#Compiling the results
res_skin <- results(dds_SS_skin_exp, alpha = 0.05)

summary(res_skin)
```

In the skin there are `r sum(res_skin$padj <= 0.05 & res_skin$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(res_skin$padj <= 0.05 & res_skin$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. Let's now save those and the non biased genes into their own objects.

```{r skin-sf-sexbiased}
#Removing the rows where padj. is NA in results
skin_res_noNA <- res_skin[!is.na(res_skin$padj),]
summary(skin_res_noNA) #We can now see that there are no outliers or 
                       #low counts since the NAs have been removed
#write.csv(as.data.frame(skin_res_noNA), "data/skin_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the skin
skin_mal_biased <- skin_res_noNA[which(skin_res_noNA$log2FoldChange >= 2
                                        & skin_res_noNA$padj <= 0.05),]
skin_fem_biased <- skin_res_noNA[which(skin_res_noNA$log2FoldChange <= -2 
                                        & skin_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the skin
skin_non_biased <- skin_res_noNA[which(skin_res_noNA$padj > 0.05),]
```

## Gonad M-F differential expression
In the sample-sample distance heatmaps and the PCA samples SSPTM3 appears to stand out quite a lot, because of this is was removed from the differential expression analysis.

```{r gonad-sf-analysis, message=FALSE}
#Create a subset of the DESeq dataset that only includes the liver samples
dds_SS_gonad <- dds_SS[, dds_SS$Organ == "Gonad"]
dds_SS_gonad <- dds_SS_gonad[, dds_SS_gonad$ID != "SSPTM3"]

#Filtering to remove any rows that have less than 10 reads total
keep <- rowSums(counts(dds_SS_gonad)) >= 10 
dds_SS_gonad <- dds_SS_gonad[keep, ]

design(dds_SS_gonad) <- ~ Sex

#Run the differential expression analysis
dds_SS_gonad_exp <- DESeq(dds_SS_gonad)

#Compiling the results
res_gonad <- results(dds_SS_gonad_exp, alpha = 0.05)

summary(res_gonad)
```

Between the testis and ovaries there are `r sum(res_gonad$padj <= 0.05 & res_gonad$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(res_gonad$padj <= 0.05 & res_gonad$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. Let's now save those and the non biased genes into their own objects.

```{r gonad-sex-biased}
#Removing the rows where padj. is NA in results
gonad_con_res_noNA <- res_gonad[!is.na(res_gonad$padj), ]
summary(gonad_con_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed
#write.csv(as.data.frame(gonad_con_res_noNA), "data/gonad_res.csv", row.names = TRUE)

#Creating an object that contains all of the male-biased and female-biased genes in the gonads
gonad_mal_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange >= 2 
                                             & gonad_con_res_noNA$padj <= 0.05),]
gonad_fem_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange <= -2 
                                             & gonad_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gonads, p>0.05
gonad_non_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$padj > 0.05),]
```

### Looking at the MA-plots for each organ

```{r MA-plot-MF, echo=FALSE, fig.height= 15, fig.width=8, fig.cap="MA-plots generated for each organ type that compares logFC to the mean expression. Female-biased and Male-biased genes are represented by color (green and purple correspondingly) and are determing with a fc cutoff of 2 and a p-value cutoff of 0.05."}
liver_MA <- ggmaplot(res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

brain_MA <- ggmaplot(res_brain,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Brain" %->% "Female Brain"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

skin_MA <- ggmaplot(res_skin,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Skin" %->% "Female Skin"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gonad_MA <- ggmaplot(res_gonad,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Testis" %->% "Ovaries"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
detach("package:PCAtools", unload = TRUE)
library(cowplot)
plot_grid(liver_MA, brain_MA, 
          skin_MA, gonad_MA,
          ncol = 1,
          nrow = 4,
          labels = c('A', 'B', 'C', 'D'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10, 4.10, 4.10))

```

From the MA plots we can see the highest number of biased genes are found in the gonads (Fig. \@ref(fig:MA-plot-MF)) and then the liver, skin, and lastly the brain The liver appears to be the only organ where there are more female-biased genes than male-biased genes.

## Investigating expression across samples in the sex-biased genes
```{r heatmap_gonads, echo=FALSE}
sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4" )

ann_colors = list(
  Sex=sex_cols)

df <- as.data.frame(SS_samples[,c("Sex", "Organ")])
rownames(df) <- SS_samples$ID

col_order_gonad <- c(
  rownames(df[df$Sex=="F"&df$Organ=="Gonad",]),
  rownames(df[df$Sex=="M"&df$Organ=="Gonad",])
)

col_order_gonad <- col_order_gonad[col_order_gonad != "SSPTM3"]

#Pull out the top 20 rows with the most expression
select_gonad_all <- order(rowMeans(counts(dds_SS_gonad_exp, 
                                          normalized = TRUE)),
                          decreasing = TRUE)[1:50]
df_gonad <- as.data.frame(colData(dds_SS_gonad)[,"Sex"])
colnames(df_gonad) <- "Sex"

#Transform the data
vsd_gonad <- vst(dds_SS_gonad_exp, blind=FALSE)

rownames(df_gonad) <- colnames(counts(dds_SS_gonad_exp, normalized = TRUE))

#Run the heat map with the function pheatmap
pheatmap(assay(vsd_gonad)[select_gonad_all,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df_gonad)

select_mal_gonad <- rownames(head(gonad_mal_biased[order(gonad_mal_biased$log2FoldChange,
                                                         decreasing = TRUE),], 
                                  n=200))

gonadM_HM <- pheatmap(assay(vsd_gonad)[select_mal_gonad, col_order_gonad], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Top 200 Male-biased genes in the Testis")

save_pheatmap_pdf(gonadM_HM, "imgs/gonadMB_heatmap.pdf",width=6,height=6)

select_fem_gonad <- rownames(head(gonad_fem_biased[order(gonad_fem_biased$log2FoldChange,
                                                         decreasing = FALSE),], 
                                  n=200))

gonadF_HM <- pheatmap(assay(vsd_gonad)[select_fem_gonad, col_order_gonad], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Top 200 Female-biased genes in the Ovaries")

save_pheatmap_pdf(gonadF_HM, "imgs/gonadFB_heatmap.pdf",width=6,height=6)

gonad_HM <- pheatmap(assay(vsd_gonad)[c(select_fem_gonad, select_mal_gonad), 
                                       col_order_gonad], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              fontsize = 12,
              annotation_legend = FALSE,
              main = "Top 200 biased genes in the Gonads")

```

```{r heatmap_skin, echo=FALSE}
col_order_skin <- c(
  rownames(df[df$Sex=="F"&df$Organ=="Skin",]),
  rownames(df[df$Sex=="M"&df$Organ=="Skin",])
)

#Pull out the top 20 rows with the most expression
select_skin <- order(rowMeans(counts(dds_SS_skin_exp, 
                                     normalized = TRUE)),
                     decreasing = TRUE)[1:50]
df_skin <- as.data.frame(colData(dds_SS_skin)[,"Sex"])
colnames(df_skin) <- "Sex"

#Transform the data
vsd_skin <- vst(dds_SS_skin_exp, blind=FALSE)

rownames(df_skin) <- colnames(counts(dds_SS_skin_exp, normalized = TRUE))

#Run the heat map with the function pheatmap
pheatmap(assay(vsd_skin)[select_skin,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df_skin)

select_mal_skin <- rownames(head(skin_mal_biased[order(skin_mal_biased$log2FoldChange,
                                                       decreasing = TRUE),], 
                                 n=200))

skinM_HM <- pheatmap(assay(vsd_skin)[select_mal_skin, col_order_skin], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Top 200 Male-biased genes in the Skin")

save_pheatmap_pdf(skinM_HM, "imgs/skinMB_heatmap.pdf",width=6,height=6)

select_fem_skin <- rownames(head(skin_fem_biased[order(skin_fem_biased$log2FoldChange,
                                                       decreasing = FALSE),], 
                                 n=200))

skinF_HM <- pheatmap(assay(vsd_skin)[select_fem_skin, col_order_skin], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Top 200 Female-biased genes in the Skin")

save_pheatmap_pdf(skinF_HM, "imgs/skinFB_heatmap.pdf",width=6,height=6)

skin_HM <- pheatmap(assay(vsd_skin)[c(select_fem_skin, select_mal_skin), 
                                       col_order_skin], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              fontsize = 12,
              annotation_legend = FALSE,
              main = "Top 200 biased genes in the Skin")
```

```{r heatmap_liver, echo=FALSE}
col_order_liver <- c(
  rownames(df[df$Sex=="F"&df$Organ=="Liver",]),
  rownames(df[df$Sex=="M"&df$Organ=="Liver",])
)

#Pull out the top 20 rows with the most expression
select_liver <- order(rowMeans(counts(dds_SS_liver_exp, 
                                     normalized = TRUE)),
                     decreasing = TRUE)[1:50]
df_liver <- as.data.frame(colData(dds_SS_liver)[,"Sex"])
colnames(df_liver) <- "Sex"

#Transform the data
vsd_liver <- vst(dds_SS_liver_exp, blind=FALSE)

rownames(df_liver) <- colnames(counts(dds_SS_liver_exp, normalized = TRUE))

#Run the heat map with the function pheatmap
pheatmap(assay(vsd_liver)[select_liver,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = TRUE, 
         annotation_col = df_liver)

select_mal_liver <- rownames(head(liver_mal_biased[order(liver_mal_biased$log2FoldChange,
                                                         decreasing = TRUE),], n=40))

liverM_HM <- pheatmap(assay(vsd_liver)[select_mal_liver, col_order_liver], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              border_color=NA,
              fontsize = 16,
              annotation_legend = FALSE,
              main = "Male-biased genes in the Liver")

save_pheatmap_pdf(liverM_HM, "imgs/liverMB_heatmap.pdf",width=6,height=6)


select_fem_liver <- rownames(head(liver_fem_biased[order(liver_fem_biased$log2FoldChange,
                                                         decreasing = FALSE),], n=200))

liverF_HM <- pheatmap(assay(vsd_liver)[select_fem_liver, col_order_liver], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              border_color=NA,
              fontsize = 16,
              annotation_legend = FALSE,
              main = "Female-biased genes in the Liver")

save_pheatmap_pdf(liverF_HM, "imgs/liverFB_heatmap.pdf",width=6,height=6)

liver_HM <- pheatmap(assay(vsd_liver)[c(select_fem_liver, select_mal_liver), 
                                       col_order_liver], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              fontsize = 12,
              annotation_legend = FALSE,
              main = "Top biased genes in the Liver")
```

```{r heatmap_brain, echo=FALSE}
col_order_brain <- c(
  rownames(df[df$Sex=="F"&df$Organ=="Brain",]),
  rownames(df[df$Sex=="M"&df$Organ=="Brain",])
)

#Pull out the top 20 rows with the most expression
select_brain <- order(rowMeans(counts(dds_SS_brain_exp, 
                                     normalized = TRUE)),
                     decreasing = TRUE)[1:50]
df_brain <- as.data.frame(colData(dds_SS_brain)[,"Sex"])
colnames(df_brain) <- "Sex"

#Transform the data
vsd_brain <- vst(dds_SS_brain_exp, blind=FALSE)

rownames(df_brain) <- colnames(counts(dds_SS_brain_exp, normalized = TRUE))

#Run the heat map with the function pheatmap
pheatmap(assay(vsd_brain)[select_brain,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df_brain)

select_mal_brain <- rownames(head(brain_mal_biased[order(brain_mal_biased$log2FoldChange,
                                                         decreasing = TRUE),], n=20))

brainM_HM <- pheatmap(assay(vsd_brain)[select_mal_brain, col_order_brain], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              border_color=NA,
              fontsize = 16,
              annotation_legend = FALSE,
              main = "Male-biased genes in the Brain")

save_pheatmap_pdf(brainM_HM, "imgs/brainMB_heatmap.pdf",width=6,height=6)

select_fem_brain <- rownames(head(brain_fem_biased[order(brain_fem_biased$log2FoldChange,
                                                         decreasing = FALSE),], n=20))

brainF_HM <- pheatmap(assay(vsd_brain)[select_fem_brain, col_order_brain], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              border_color=NA,
              fontsize = 16,
              annotation_legend = FALSE,
              main = "Female-biased genes in the Brain")

save_pheatmap_pdf(brainF_HM, "imgs/brainFB_heatmap.pdf",width=6,height=6)

brain_HM <- pheatmap(assay(vsd_brain)[c(select_fem_brain, select_mal_brain), 
                                       col_order_brain], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df[1],
              annotation_colors=ann_colors,
              fontsize = 16,
              border_color=NA,
              annotation_legend = FALSE,
              main = "Top biased genes in the Brain")
```

```{r}
mbgonad <- image_ggplot(image_read_pdf('imgs/gonadMB_heatmap.pdf'),
                        interpolate = TRUE)
fbgonad <- image_ggplot(image_read_pdf('imgs/gonadFB_heatmap.pdf'),
                        interpolate = TRUE)
mbskin <- image_ggplot(image_read_pdf('imgs/skinMB_heatmap.pdf'),
                        interpolate = TRUE)
fbskin <- image_ggplot(image_read_pdf('imgs/skinFB_heatmap.pdf'),
                        interpolate = TRUE)
mbliver <- image_ggplot(image_read_pdf('imgs/liverMB_heatmap.pdf'),
                        interpolate = TRUE)
fbliver <- image_ggplot(image_read_pdf('imgs/liverFB_heatmap.pdf'),
                        interpolate = TRUE)
mbbrain <- image_ggplot(image_read_pdf('imgs/brainMB_heatmap.pdf'),
                        interpolate = TRUE)
fbbrain <- image_ggplot(image_read_pdf('imgs/brainFB_heatmap.pdf'),
                        interpolate = TRUE)
figHM <- wrap_plots(
  mbgonad,
  fbgonad,
  mbskin,
  fbskin,
  mbliver,
  fbliver,
  mbbrain,
  fbbrain,
 ncol= 2,
 nrow = 4
)
figHM <- figHM + plot_annotation(tag_levels = 'A',
                                 theme = theme(plot.tag = element_text(size = 12))) 

ggsave("imgs/heatmapSup.pdf",figHM, height=8, width=5)
ggsave("imgs/heatmapSup.png",figHM, height=8, width=5) # also save as a png
```

## Looking at shared sex-biased genes with Upset plots
I now want to see if there are genes that are female/male-biased in multiple organs or if there are genes that are female-biased in one organ and then male-biased in a different organ, etc.. In order to do this I will be creating upset plots first for the sexes individually and then across all biased genes.

```{r upset-plot-mal, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in males."}
#Male-biased genes
listInputMB <- list("Testis" = rownames(gonad_mal_biased),
                  "Male Brain"=rownames(brain_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased),
                  "Male Skin"=rownames(skin_mal_biased)
                  )
upset(fromList(listInputMB),
      mainbar.y.label = "Number of Shared Male-Biased Genes",
      sets.x.label = "Total Number of Male-Biased Genes",
      point.size = 3)
```

```{r upset-plot-fem, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in females."}
#Female
listInputFB <- list("Ovaries" = rownames(gonad_fem_biased),
                  "Female Brain"=rownames(brain_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased),
                  "Female Skin"=rownames(skin_fem_biased)
                  )
upset(fromList(listInputFB),
      mainbar.y.label = "Number of Shared Female-Biased Genes",
      sets.x.label = "Total Number of Female-Biased Genes",
      point.size = 3)
```

```{r upset-plot-all, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs across both sexes."}
#Esporting data used to make the Upset plot (run once and them commented out)
#write.table(as.data.frame(gonad_mal_biased), "data/gonad_mal_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(gonad_fem_biased), "data/gonad_fem_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(brain_mal_biased), "data/brain_mal_biased.txt", 
 #            row.names = TRUE)
#write.table(as.data.frame(brain_fem_biased), "data/brain_fem_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(liver_mal_biased), "data/liver_mal_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(liver_fem_biased), "data/liver_fem_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(skin_mal_biased), "data/skin_mal_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(skin_fem_biased), "data/skin_fem_biased.txt", 
 #           row.names = TRUE)

#Checking there is no overlap between the two sexes
listInputall <- list("Testis" = rownames(gonad_mal_biased),
                  "Male Brain"=rownames(brain_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased),
                  "Male Skin"=rownames(skin_mal_biased),
                  "Ovaries" = rownames(gonad_fem_biased),
                  "Female Brain"=rownames(brain_fem_biased),
                  "Female Skin"=rownames(skin_fem_biased),
                  "Female Liver"=rownames(liver_fem_biased))

upset(fromList(listInputall),
      mainbar.y.label = "Number of Shared Sex-Biased Genes",
      sets.x.label = "Total Number of Sex-Biased Genes",
      point.size = 3,
      nsets = 8,
      nintersects = NA)

```

In both males and females the highest number of shared sex-biased genes is found between the gonads and the skin (Fig. \@ref(fig:upset-plot-mal), \@ref(fig:upset-plot-fem)) followed by the gonads and the liver. There are no genes that are shared across the same organ between males and females (i.e. biased in males and females in the liver) which is good (Fig. \@ref(fig:upset-plot-all)). There are no genes that are biased in all four organs, but for females there are 2 genes that are female-biased in the liver, skin, and ovaries, and in males there is 1 gene that is biased in the liver, skin, and testis (Fig. \@ref(fig:upset-plot-mal), \@ref(fig:upset-plot-fem)).

## Looking at variation in logFC across both sex-bias categories and tissue types
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset.

```{r logFC-long-dataset}
logFC_long<- data.frame(
  tissue=c(rep("Brain",nrow(brain_fem_biased)),
           rep("Brain", nrow(brain_mal_biased)),
           rep("Brain", nrow(brain_non_biased)),
           rep("Gonad", nrow(gonad_fem_biased)),
           rep("Gonad", nrow(gonad_mal_biased)),
           rep("Gonad", nrow(gonad_non_biased)),
           rep("Liver", nrow(liver_fem_biased)),
           rep("Liver", nrow(liver_mal_biased)),
           rep("Liver", nrow(liver_non_biased)),
           rep("Skin", nrow(skin_fem_biased)),
           rep("Skin", nrow(skin_mal_biased)),
           rep("Skin", nrow(skin_non_biased))
         ),
  bias=c(rep("FB",nrow(brain_fem_biased)),
         rep("MB",nrow(brain_mal_biased)),
         rep("NB", nrow(brain_non_biased)),
         rep("FB", nrow(gonad_fem_biased)),
         rep("MB", nrow(gonad_mal_biased)),
         rep("NB", nrow(gonad_non_biased)),
         rep("FB", nrow(liver_fem_biased)),
         rep("MB", nrow(liver_mal_biased)),
         rep("NB", nrow(liver_non_biased)),
         rep("FB", nrow(skin_fem_biased)),
         rep("MB", nrow(skin_mal_biased)),
         rep("NB", nrow(skin_non_biased))
         ),
  logFC=c(brain_fem_biased$log2FoldChange,
          brain_mal_biased$log2FoldChange,
          brain_non_biased$log2FoldChange,
          gonad_fem_biased$log2FoldChange,
          gonad_mal_biased$log2FoldChange,
          gonad_non_biased$log2FoldChange,
          liver_fem_biased$log2FoldChange,
          liver_mal_biased$log2FoldChange,
          liver_non_biased$log2FoldChange,
          skin_fem_biased$log2FoldChange,
          skin_mal_biased$log2FoldChange,
          skin_non_biased$log2FoldChange
          ),
  geneID=c(rownames(brain_fem_biased),
           rownames(brain_mal_biased),
           rownames(brain_non_biased),
           rownames(gonad_fem_biased),
           rownames(gonad_mal_biased),
           rownames(gonad_non_biased),
           rownames(liver_fem_biased),
           rownames(liver_mal_biased),
           rownames(liver_non_biased),
           rownames(skin_fem_biased),
           rownames(skin_mal_biased),
           rownames(skin_non_biased)
           )
  
)

#Exporting this data, run once and then commented out
#write.csv(logFC_long, "data/logFC_long_sexbias.csv", row.names = FALSE)
```

After creating the long version of the dataset I can generate a plot to demonstrate the variation in logFC across bias and tissue type.

```{r FC-var-plot, echo=FALSE, fig.cap="Absolute value of the logFC for genes that are female-biased, male-biased, and unbiased across each tissue type. Raw fold-change values are added ontop of the boxplot as jitters."}

my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "NB" = "darkgray")

ggplot(logFC_long,
       aes(x = bias, y = abs(logFC), fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_point(data=logFC_long, aes(x=bias, y=abs(logFC)),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y="|logFC|") +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

```

There appears to be differences in the logFC across both the organs and the sex-bias groups (male-biased, fem-biased, and non-biased) (Fig. \@ref(fig:FC-var-plot)). In the brain, potentially skin, and especially in the liver, the female-biased genes are more biased (i.e. higher logFC) than the male-biased genes. This reverses in the gonad. In every case the non-biased genes have the lowest median logFC even though there is some variation.

I am going to fit a linear model to the data.

```{r FC-var-stats, warning=FALSE, eval=FALSE}
#Looking at some summary statistics for logFC between the different groups
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), median)
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), sd)

model <- lm(abs(logFC_long$logFC)~ logFC_long$tissue * logFC_long$bias)
summary(model)
```

## Categorizing sex-specific genes
Now that we have investigated what the sex-bias is like across the different tissue types I want to dive further into genes with **sex-specific expression**. These are genes expressed only in one sex or the other. I am classifying a sex-specific genes within each tissue as ones where the expression of that gene is less than 10 for all of one sex and there is a median of $\ge 20$ in the other sex. 

Prior to running the for loop, any "outliers" that were determined by DESeq for each of the differential expression analyses (within the organs) were removed from the pool of genes. Additionally, normalized counts were used for the classification of sex-specific genes as DESeq2 uses the moralized counts for all of the logFC/sexbias calculations.

```{r sex-specific-for-loop}
#Store all of the single factor analysis outputs in a list
DE_models <- list(dds_SS_brain_exp, 
                  dds_SS_gonad_exp, 
                  dds_SS_liver_exp, 
                  dds_SS_skin_exp)

#Create an empty list to store my sex-specific datasets in
SS_sex_specific_genes <- list()

#For each DESeq dataset in DE_models pull out the Mspecific and Fspecific
#genes based on medians
for (dataset in DE_models) {
  
  #Pull out the geneIDs for genes that were categorized as "outliers" by DESeq2
  ##Calculate the Cooks threshold that would have been used
  np <- length(resultsNames(dataset))
  nsamp <- ncol(dataset)
  cooks_thresh <- qf(0.99, df1 = np, df2 = nsamp - np)
  
  ##Apply threshold calculated above to the cooks values in the dataset
  out_ids <- names(mcols(dataset)$maxCooks[mcols(dataset)$maxCooks >
                                                cooks_thresh])
  
  #Filtering the dds dataset to remove the outliers identified by DESeq
  dataset_filtered <- dataset[!(rownames(dataset) %in% out_ids), ]
  
  
  #Male-specific Genes
  ##Pull out all of the rows where fem count <=10 in every female sample
  fem10_organ_names <- which(rowSums(t(apply(counts(dataset_filtered,
                                                    normalized = TRUE)[, dataset_filtered$Sex == "F"],
                                             1,
                                             function(x) x <= 10)
                                       )
                                     ) == ncol(counts(dataset_filtered, 
                                                      normalized = TRUE)[, dataset_filtered$Sex == "F"])
                             )
  
  fem10_organ <- counts(dataset_filtered,
                        normalized = TRUE)[rownames(counts(dataset_filtered,
                                                          normalized = TRUE)) %in%
                                            names(fem10_organ_names),]
  
  ##Pull out the rows where median of male count >=50  
  mal50_organ <- apply(counts(dataset_filtered,
                              normalized = TRUE)[, dataset_filtered$Sex == "M"],
                       1,
                       function(x) median(x) >= 50)
  
  ##Keep only rows where all female samples <=10 and the male median >= 50
  fem10_mal50_organ <- fem10_organ[rownames(fem10_organ) %in%
                                     names(mal50_organ[mal50_organ == TRUE]),
                                   ]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", dataset$Organ[1])
  SS_sex_specific_genes[[organ_malsp]] <- fem10_mal50_organ
  
  
  #Female-specific Genes
  ##Pull out ll the rows where male count is <=10 in every male sample
  mal10_organ_names <- which(rowSums(t(apply(counts(dataset_filtered,
                                                    normalized = TRUE)[, dataset_filtered$Sex == "M"],
                                             1,
                                             function(x) x <= 10)
                                       )
                                     ) == ncol(counts(dataset_filtered,
                                                      normalized = TRUE)[, dataset_filtered$Sex == "M"])
                             )
  
  mal10_organ <- counts(dataset_filtered,
                        normalized = TRUE)[rownames(counts(dataset_filtered,
                                                           normalized = TRUE)) %in%
                                             names(mal10_organ_names), ]
  
  ##Pull out rows where median of female count is >=50 
  fem50_organ <- apply(counts(dataset_filtered,
                              normalized = TRUE)[, dataset_filtered$Sex == "F"],
                       1,
                       function(x) median(x) >= 50)
  
  ##keep only the rows where male <=10 and fem median >= 50
  mal10_fem50_organ <- mal10_organ[rownames(mal10_organ) %in%
                                     names(fem50_organ[fem50_organ == TRUE]),
                                   ]
  
  ##Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", dataset$Organ[1])
  SS_sex_specific_genes[[organ_femsp]] <- mal10_fem50_organ
  
}
```

### Investigating the results of the sex-specific subsetting
Let's now take a look at how many sex-specific genes we have in each tissue type:
```{r sex-specific-counts, echo=FALSE, message=FALSE, warning=FALSE, eval =FALSE}
sex_specific_counts <- data.frame(tissue = rep(c("Brain", "Gonad", "Liver", "Skin"), 2),
                                  sex = c(rep("M", 4),
                                          rep("F", 4)),
                                  num_genes = c(nrow(SS_sex_specific_genes$Brain_male_specific),
                                                nrow(SS_sex_specific_genes$Gonad_male_specific),
                                                nrow(SS_sex_specific_genes$Liver_male_specific),
                                                nrow(SS_sex_specific_genes$Skin_male_specific),
                                                nrow(SS_sex_specific_genes$Brain_female_specific),
                                                nrow(SS_sex_specific_genes$Gonad_female_specific),
                                                nrow(SS_sex_specific_genes$Liver_female_specific),
                                                nrow(SS_sex_specific_genes$Skin_female_specific))
                                  )

sex_specific_counts %>% 
  group_by(tissue, sex) %>% 
  summarise("Number of Sex-specific Genes" = sum(num_genes)) %>%
  kable(caption = "Sex-specific Genes", format = "html",
        col.names = c("Tissue Type", "Sex", "Number of Sex-specific Genes")) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F, position = "left")
```

### Pulling out the gene IDs for all of the sex-specific genes
```{r SS-gene_IDs, eval=FALSE}
#Run once to create the file and then commented out
write.table(cbind(rownames(SS_sex_specific_genes$Brain_male_specific)),
            'data/SS_malB_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(SS_sex_specific_genes$Brain_female_specific)),
            'data/SS_femB_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(SS_sex_specific_genes$Gonad_male_specific)),
            'data/SS_malGon_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(SS_sex_specific_genes$Gonad_female_specific)),
            'data/SS_femGon_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(SS_sex_specific_genes$Liver_female_specific)),
            'data/SS_femL_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(SS_sex_specific_genes$Skin_male_specific)),
            'data/SS_malS_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(SS_sex_specific_genes$Skin_female_specific)),
            'data/SS_femS_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

```

## Creating categories and binning the sex-biased genes based on degree of logFC
This binning was done on all of the sex-biased genes, unbiased genes were given the category of "unbiased". The cutoffs for the different binning categories are as follows:

    1. Low - biased = LFC 2 - 3
    2. Medium - biased = LFC 3 - 5
    3. High - biased = LFC 5 - 10
    4. Extreme sex bias = LFC > 10

Counts of the sex-specific genes will be added on separately as they were not classified based on fold-change but rather a presence in one sex and an absence in the other sex. A for loop will then be used to make sure genes are not counted twice (i.e. as both extremely biased and sex-specific).

```{r bin-sex-bias, fig.cap="Counts of sex-biased genes that fall into the different categorties for sex-bias in males and females across organs."}
#Make a vector that contains all of the groupings
biased_bins <- c("Unbiased", "Low", "Med", "High", "Extreme", "Sex-specific")

#Create a new column in the dataset and use ifelse statements to set the category limits
#abs(logFC) was used to account for the fem-biased genes possessing negative values
logFC_long$bias_cat <- ifelse(logFC_long$bias == "NB",
                              biased_bins[1],
                              ifelse(abs(logFC_long$logFC) >= 2 & abs(logFC_long$logFC) < 3,
                                     biased_bins[2],
                                     ifelse(abs(logFC_long$logFC) >= 3 & abs(logFC_long$logFC) < 5,
                                            biased_bins[3],
                                            ifelse(abs(logFC_long$logFC) >= 5 & abs(logFC_long$logFC) < 10,
                                                   biased_bins[4],
                                                   biased_bins[5])
                                            )
                                     )
                              )

#Making sure the genes we categorized as sex-specific are labeled as sex-specific for their 
#bias cat. in the dataset
organs <- c("Brain", "Brain", "Gonad", "Gonad", "Liver", "Liver", "Skin", "Skin")
bias <- c("MB", "FB", "MB", "FB", "MB", "FB", "MB", "FB")

for(i in 1:length(SS_sex_specific_genes)){

  tmp <- SS_sex_specific_genes[[i]]
  tmp <- as.data.frame(tmp)
  tmp$geneID <- rownames(tmp)
  
  if(nrow(tmp) > 0){
  
  for(j in 1:nrow(tmp)){
    
    one_gene <- tmp[j, ]
   
    if(one_gene[["geneID"]] %in%
       logFC_long[logFC_long$tissue == organs[[i]] &
                  logFC_long$bias == bias[[i]],"geneID"]){
       
      logFC_long[logFC_long$geneID == one_gene[["geneID"]] &
                   logFC_long$tissue == organs[[i]] &
                   logFC_long$bias == bias[[i]],
                 "bias_cat"] <- "Sex-specific"
    }else{
      
      one_gene_dat <- data.frame(matrix(ncol= ncol(logFC_long),
                                        nrow=1))
      colnames(one_gene_dat) <- colnames(logFC_long)
      
      one_gene_dat$tissue <- organs[[i]]
      one_gene_dat$geneID <- one_gene[["geneID"]]
      one_gene_dat$bias <- bias[[i]]
      one_gene_dat$bias_cat <- "Sex-specific"
      rownames(one_gene_dat) <- NULL
      
      logFC_long <- rbind(one_gene_dat, logFC_long)
      
      rownames(logFC_long) <- NULL
    }
  }}else{
    
  }
 }

#Exporting this data, run once and then commented out
#write.csv(logFC_long, "data/logFC_long_bias_cats.csv", row.names = FALSE)

#Create  subset of our long dataset that does not include the non-biased genes
logFC_long_noNB <- logFC_long[logFC_long$bias_cat != "Unbiased",]

#Make a table to count the number of genes in each category for each organ
bias_cat_table <- table(logFC_long_noNB$bias, logFC_long_noNB$bias_cat, logFC_long_noNB$tissue)

#Add the counts of sex-specific genes to the table
bias_cat_brain <- bias_cat_table[,, "Brain"]
#write.table(bias_cat_brain, "data/bias_cat_brain.txt", row.names = TRUE)

bias_cat_gonad <- bias_cat_table[,, "Gonad"]
#write.table(bias_cat_gonad, "data/bias_cat_gonad.txt", row.names = TRUE)

bias_cat_liver <- bias_cat_table[,, "Liver"]
#write.table(bias_cat_liver, "data/bias_cat_liver.txt", row.names = TRUE)

bias_cat_skin <- bias_cat_table[,, "Skin"]
#write.table(bias_cat_skin, "data/bias_cat_skin.txt", row.names = TRUE)

#Plot the counts for each tissue type
par(mfrow=c(1, 4))
barplot(bias_cat_brain[,biased_bins[biased_bins!="Unbiased"]], 
        beside = TRUE, 
        ylim = c(0, max(bias_cat_brain) + 10), 
        col = c("#7fc97f", "#beaed4"),
        main = "Brain")

barplot(bias_cat_gonad[,biased_bins[biased_bins!="Unbiased"]], 
        beside = TRUE, 
        ylim = c(0, max(bias_cat_gonad) + 10), 
        col = c("#7fc97f", "#beaed4"),
        main = "Gonad")

barplot(bias_cat_liver[,biased_bins[biased_bins!="Unbiased"]], 
        beside = TRUE, 
        ylim = c(0, max(bias_cat_liver) + 10), 
        col = c("#7fc97f", "#beaed4"),
        main = "Liver")

barplot(bias_cat_skin[,biased_bins[biased_bins!="Unbiased"]], 
        beside = TRUE, 
        ylim = c(0, max(bias_cat_skin) + 10), 
        col = c("#7fc97f", "#beaed4"),
        main = "Skin")
legend("topright", legend = c("Female-biased", "Male-biased"), fill = c("#7fc97f", "#beaed4"))

```

# Gene Ontologogy Analysis and BLASTing the different genes
There are several things that I want to accomplish with the gene ontology analysis, to start with I want to classify the sex-biased genes and then the sex-specific genes.

## BLASTING against the zebrafish
The first step to the GO analysis is to BLAST the trinity sequences against one of the PANTHER organisms. The zebrafish, _Danio reiro_, was used for this as it was the only fish in the list.

A BLAST database was generated with the _D. reiro_ proteome as:
`makeblastdb -in ../ncbi_dataset/data/GCF_000002035.6/protein.faa -out d_rerio_prot -dbtyp prot`

The trinity gene IDs that correspond to ALL of the female- or male-biased genes were pulled out and saved to `.txt` files.

```{r biased-trin-ids, eval=FALSE}
write.table(cbind(rownames(brain_fem_biased)),
            'SS_femB_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(brain_mal_biased)),
            'SS_malB_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(gonad_fem_biased)),
            'SS_femGon_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(gonad_mal_biased)),
            'SS_malGon_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(liver_fem_biased)),
            'SS_femL_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(liver_mal_biased)),
            'SS_malL_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(skin_fem_biased)),
            'SS_femS_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(skin_mal_biased)),
            'SS_malS_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```

The `.txt` files were imported into the RCC and then BLASTed against the zebrafish proteome using the following script:

```{bash blast_pipeline, file = 'bash/blast_pipeline.sh', eval = FALSE}


```

This script was run as `bash bash_scripts/blast_pipeline.sh scovelli_high_exp_genes/high_expTG_names other_scripts/subset_fasta_file trinity_supertran_scovelli.fasta blastx ../genomes/ scovelli_high_exp_genes/high_exp_results/`.

The trinity gene IDs that correspond to all of the sex-specific genes were pulled out above. Those were BLASTed with the script above using **blastn** against the _Syngnathus scovelli_ genome as: 
`bash bash_scripts/blast_pipeline.sh scovelli_high_exp_genes/high_expTG_names other_scripts/subset_fasta_file trinity_supertran_scovelli.fasta blastn ../genomes/s_scov/s_scov_genome _blast.txt scovelli_high_exp_genes/high_exp_results/`.

## Read in and filter the BLAST results
I then exported all of the blast results as .txt files from the RCC and need to first read them into R so that I can filter the results and make sure only one hit is kept per gene.

```{r readin_blast_results_dreiro, echo=FALSE}
#Specify the directory where BLAST results are located
blast_path <- "data/scovelli_BLAST/"

#Create a list of the files I want
SS_blast_files <- list.files(blast_path, pattern = "reiro")

#Create an empty list to store my datasets in
SS_blast_list <- list()

#Create a loop to read in all of the blast results
for(file_name in SS_blast_files){
  
  #Read the file and store it in a new object
  file_data <- read.delim(file.path(blast_path, file_name), stringsAsFactors = FALSE, quote = "", sep = "\t", header = FALSE)

  #Add column names to the dataset
  colnames(file_data) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "reiro_prot_info", "prot_id", "reiro_prot_start", "reiro_prot_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Create a new object with a name based on the file name
  blast_name <- sub(".txt$", "", file_name) #Removes the file extension
  SS_blast_list[[blast_name]] <- file_data
}

```

```{r readin_blast_results_SSgenes, echo=FALSE}
#Create a list of the files I want
SS_blast_files_sex_specific <- list.files(blast_path, pattern = "specific")

#Create an empty list to store my datasets in
SS_blast_list_sex_specific <- list()

#Create a loop to read in all of the blast results
for(file_name in SS_blast_files_sex_specific){
  
  #Read the file and store it in a new object
  file_data <- read.delim(file.path(blast_path, file_name), 
                          stringsAsFactors = FALSE, quote = "", 
                          sep = "\t", header = FALSE)

  #Add column names to the dataset
  colnames(file_data) <- c("trin_geneid", 
                           "trin_gene_start", "trin_gene_end", 
                           "scov_gene_info", "scov_prot_info",
                           "scov_gene_start", "scov_gene_end", 
                           "evalue", "bit_score", "length", "pident", 
                           "gaps")
  
  #Create a new object with a name based on the file name
  blast_name <- sub(".txt$", "", file_name) #Removes the file extension
  SS_blast_list_sex_specific[[blast_name]] <- file_data
}

```

Not that I have all of the BLAST files for SS genes and SB genes in R I am going to filter them:

```{r filter-BLAST-dreiro}
#Use lapply to apply the function to each dataset stored in the list created above
blast_output_filtered <- lapply(SS_blast_list, function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "reiro_prot_info", "prot_id", "reiro_prot_start", "reiro_prot_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID$trin_geneid){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- subset(data_frame, trin_geneid==gene)
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue == min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident == max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)

    # check that there is a single gene ID in scovelli, if not, only first row is kept
    if(length(gsub("^.*(GeneID:\\d+)\\].*$","\\1",uni_gene_subset_lowe_highpid$reiro_prot_info))>1){
    
    
      uni_gene_subset_lowe_highpid<-uni_gene_subset_lowe_highpid[1,]
    }
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
})

```

```{r filter-BLAST-SS}
#Use lapply to apply the function to each dataset stored in the list created above
blast_output_filtered_SS <- lapply(SS_blast_list_sex_specific, function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <- c("trin_geneid", "trin_gene_start", "trin_gene_end",
                        "scov_gene_info", "scov_prot_info",
                        "scov_gene_start", "scov_gene_end", 
                        "evalue", "bit_score", "length", "pident", "gaps")
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID$trin_geneid){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- subset(data_frame, trin_geneid==gene)
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue ==
                                                   min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- 
      uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident ==
                                   max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)

    # check that there is a single gene ID in scovelli, if not, only first row is kept
    if(length(gsub("^.*(GeneID:\\d+)\\].*$",
                   "\\1",
                   uni_gene_subset_lowe_highpid$scov_gene_info))>1){
      
      uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe_highpid[1,]
    }
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
})

```

## Collect gene names corresponding to each protein -sex-biased genes
In order to run PANTHER we need IDs that follow one of their supported formats. From NCBI, the gene name works (e.g. rpl27). That information is not automatically supplied with the BLAST output, but we can use the .gff file from NCBI to pull out the gene names that correspond to the proteinID (e.g. NP_956018.1).

```{r gff-gene-names}
#Read in the GFF file for zebrafish that has info about the geneID
reiro_gff <- read.delim("data/d_reiro_genomic.gff",
                        header = FALSE,
                        comment.char = "#")

#Keep only the columns I want and rename them
reiro_gff <- reiro_gff[,c(1:5,9)]
colnames(reiro_gff) <- c("seqname", "source", "feature", "start", "end", "gene_info")

#Subset the dataset to only include the rows we're interested in
genes <- reiro_gff[reiro_gff$feature == "CDS",]

#Pull out the gene name that corresponds to each protein ID
genes$gene_name <- gsub("^(.*;)(gene=)(\\w+\\d*);(.*$)", "\\3", genes$gene_info)
genes$prot_id <- gsub("^(.*;)(protein_id=)(.*)$", "\\3", genes$gene_info)

#Merge the gene names pulled out above with the rest of the BLAST datasets based on the proteinID
blast_output_merged <- lapply(blast_output_filtered, function(dataframe){
  
   output <- merge(dataframe, unique(genes[,7:8]), by = "prot_id")
  
   return(output)
})

```

I now need to save only the gene_id for all of the BLASTed genes to then upload to PANTHER.

```{r export-danio-geneID, eval=FALSE}

write.table(c(blast_output_merged$SS_femB_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$SS_femGon_biased_TRgenes_reiro_blast$gene_name,
              blast_output_merged$SS_femL_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$SS_femS_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$SS_malB_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$SS_malGon_biased_TRgenes_reiro_blast$gene_name,
              blast_output_merged$SS_malL_biased_TRgenes_reiro_blast$gene_name,
              blast_output_merged$SS_malS_biased_TRgenes_reiro_blast$gene_name),
           'SS_GOnames_SBG.txt', 
            sep = " ", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

```

## Pull out the gene names and protein ID for sex-specific BLAST
Once all the BLAST results for the sex-specific genes have been filtered I want to pull out specifically the name of the gene and the protein ID from the  `scov_gene_info` column and save them as their own column. 

```{r scov-geneIDs}
for (i in 1:length(blast_output_filtered_SS)) {
  
  #Pulling out the gene name and adding it to a new column
  blast_output_filtered_SS[[i]]$gene <- gsub("^(.*)(gene=)(\\w+\\d*)(.*$)",
                                             "\\3",
                                             blast_output_filtered_SS[[i]]$scov_gene_info)
  
  #Pulling out the protein name and adding it to a new column
  blast_output_filtered_SS[[i]]$protein <- gsub("^(.*)(protein=)(.*)([[:graph:]]\\s[[:graph:]])(protein_id=)(.*$)",
                                                "\\3",
                                                blast_output_filtered_SS[[i]]$scov_gene_info)
  
  
}

```

Now that I have isolated the gene names and protein IDS I want to reform the BLAST results so that they can all be compiled into one single dataset.

```{r long_BLASTSS}
#Create a long format dataset that has the tissue information, BLAST gene and protein, 
#geneID, and sex 
blast_long_SS <- data.frame(
  tissue=c(rep("Brain", 
               sum(nrow(blast_output_filtered_SS$SS_femB_specific_TRgenes_blast),
                   nrow(blast_output_filtered_SS$SS_malB_specific_TRgenes_blast))),
           rep("Gonad", 
               sum(nrow(blast_output_filtered_SS$SS_femGon_specific_TRgenes_blast),
                   nrow(blast_output_filtered_SS$SS_malGon_specific_TRgenes_blast))),
           rep("Liver",
               sum(nrow(blast_output_filtered_SS$SS_femL_specific_TRgenes_blast))),
           rep("Skin",
               sum(nrow(blast_output_filtered_SS$SS_femS_specific_TRgenes_blast),
                   nrow(blast_output_filtered_SS$SS_malS_specific_TRgenes_blast)))
         ),
  sex=c(rep("Female",
            nrow(blast_output_filtered_SS$SS_femB_specific_TRgenes_blast)),
        rep("Male",
            nrow(blast_output_filtered_SS$SS_malB_specific_TRgenes_blast)),
        rep("Female",
            nrow(blast_output_filtered_SS$SS_femGon_specific_TRgenes_blast)),
        rep("Male",
            nrow(blast_output_filtered_SS$SS_malGon_specific_TRgenes_blast)),
        rep("Female",
            nrow(blast_output_filtered_SS$SS_femL_specific_TRgenes_blast)),
        rep("Female",
            nrow(blast_output_filtered_SS$SS_femS_specific_TRgenes_blast)),
        rep("Male",
            nrow(blast_output_filtered_SS$SS_malS_specific_TRgenes_blast))
        ),
  gene=c(blast_output_filtered_SS$SS_femB_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$SS_malB_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$SS_femGon_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$SS_malGon_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$SS_femL_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$SS_femS_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$SS_malS_specific_TRgenes_blast$gene
         ),
  protein=c(blast_output_filtered_SS$SS_femB_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$SS_malB_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$SS_femGon_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$SS_malGon_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$SS_femL_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$SS_femS_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$SS_malS_specific_TRgenes_blast$protein
            ),
  trinityID=c(blast_output_filtered_SS$SS_femB_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$SS_malB_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$SS_femGon_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$SS_malGon_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$SS_femL_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$SS_femS_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$SS_malS_specific_TRgenes_blast$trin_geneid)
  )

#write.csv(blast_long_SS, "SS_BLAST_results_sex_specific.csv", row.names = FALSE)
```

## Reading in the PANTHER results - Sex-biased Genes
In PANTHER I uploaded the `SS_GOnames_SBG.txt` file, selected _Danio rerio_ as the organism, and then chose "Functional classification viewed in gene list" as the analysis type. 

I exported those results to a .txt file and will now be reading that into R and visualizing the results.

```{r panther-results-SBG}
#Read in the .txt file containing PANTHER results
panther <- read.delim("data/pantherGeneList_SBG.txt", 
                      header = FALSE)

#Add the approporiate column names for the PANTHER data
colnames(panther) <- c("GeneID", "MappedID", 
                       "GeneName", "pantherFAM", 
                       "panther_prot_class", "species")

#Add the identified PANTHER protein classes to the BLAST datasets
blast_output_merged$SS_femB_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$SS_femB_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$SS_femGon_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$SS_femGon_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$SS_femL_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$SS_femL_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$SS_femS_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$SS_femS_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$SS_malB_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$SS_malB_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$SS_malGon_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$SS_malGon_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$SS_malL_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$SS_malL_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$SS_malS_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$SS_malS_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)

```

Now that we have the PANTHER protein classes attached to all of the successfully BLASTed sex-biased genes within each organ, we can plot the proportions of genes falling into each category to see what is most important for each of the organ types.

```{r plot-panther-SBG}
#Count the number of genes that are falling into each protein class
go_tables <- lapply(blast_output_merged, function(dat){
  
  #Create a table that counts the number of genes falling within each PANTHER prot. class
  panther_table <- data.frame(table(dat$panther_prot_class))
  
  #Make sure the protein names are in there as a character
  panther_table$Var1 <- as.character(panther_table$Var1)
  
  #Change all of the null results to say Unclassified
  panther_table$Var1[panther_table$Var1==""] <- "Unclassified"

  return(panther_table)
  })

#Merge all of the different organ data into one dataset
#Create an empty dataframe to store all of the values in
all_go_dat <- data.frame(matrix(ncol=4,nrow=0))

#Add column names to that empty dataset
colnames(all_go_dat) <- c(
  colnames(go_tables$SS_femB_biased_TRgenes_reiro_blast),
  "bias_cat", "tissue")

#Create a vector of you tissues, in the same order that they appear in the go_table list
tissues <- c("Brain", "Gonad", "Liver", "Skin",
             "Brain", "Gonad", "Liver", "Skin")

#Loop through all of the go_tables and combine into one
for(i in 1:length(go_tables)){
  #browser()
  tmp <- go_tables[[i]]
  if (nrow(tmp) == 0) {
    print(tmp)
  }else{
  
  tmp$bias_cat <- names(go_tables)[i]
  tmp$tissue <- tissues[[i]]
  all_go_dat <- rbind(all_go_dat, tmp)}
  
}

#Calculate the frequency at which each protein class occurs within the different tissue types
##Create an empty dataset to store the values in
all_go_sums <- data.frame(matrix(ncol = 4,
                                 nrow = 0))

##Add appropriate column names
colnames(all_go_sums) <- c("Freq", "prot_class", 
                           "tissue", "prop")

##Loop through each organ type to calculate frequencies
for(organ in unique(all_go_dat$tissue)){
  
  #subset the dataset based on the organ
  tmp <- all_go_dat[all_go_dat$tissue == organ, ]
  
  #Sum up all of the genes in each category
  go_sums <- as.data.frame(tapply(tmp$Freq, tmp$Var1, sum))
  
  #Add the protein class as a column rather than row names
  go_sums$prot_class <- rownames(go_sums)
  
  #Change the column name to "Frequency"
  colnames(go_sums)[1] <- "Freq"
  
  #Add the organ in a separate column and remove rownames
  go_sums$tissue <- organ
  rownames(go_sums) <- NULL
  
  #Calculate the respective proportion for each protein class
  go_sums$prop <- go_sums$Freq/sum(go_sums$Freq)
  
  #Export the data to the previously created dataframe
  all_go_sums <- rbind(all_go_sums, go_sums)
}

#Change all of the protein classes that have a frequency of less than 0.02 to "Other"
for(class in unique(all_go_sums$prot_class)){
  
  #Pull out the rows that contain that protein class
  match_rows <- which(all_go_sums$prot_class==class)
  
  #If any of the proportions associated with those rows is less than 0.02, classify as other
  if(any(all_go_sums$prop[match_rows]>0.02)==FALSE){
    
    all_go_sums$prot_class[match_rows] <- " other"
    
  }
}

#Change format of "Unclassified so it will come first"
all_go_sums$prot_class[all_go_sums$prot_class=="Unclassified"] <-  " unclassified"

#write.csv(all_go_sums, "data/GO_freq_SBG.csv", row.names = FALSE)

#Generate the GO plot
organ_cols <- c("Brain" = "#6E8B3D", "Gonad" = "#EEB422",
                "Liver" = "#EE8262", "Skin" = "#6959CD")

ggplot(all_go_sums, 
       aes(fct_rev(prot_class), prop, fill = tissue)) +   
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = organ_cols) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=20),
        axis.text.y = element_text(size=10),
        text=element_text(size=20),
        legend.position = "bottom") +
  coord_flip() + 
  labs(y="Proportion of sex-biased genes", x="")
```

## Running the gene ontology analysis for the GONADS only
I want to look further into differences between the testis and the ovaries in terms of what GO terms are represented in the sex-biased genes and to also conduct an over representation analysis for them individually. To do that I am going to walk through some of the same steps highlighted above, but now just for the testis and ovaries.

I am starting with reading in just the gonad data:

```{r readin_blast_results_dreiro_gonads, echo=FALSE}
#Create a list of the files I want
SS_blast_files_gonads <- list.files(blast_path, 
                                    pattern = "Gon_biased")

#Create an empty list to store my datasets in
SS_blast_list_gonads <- list()

#Create a loop to read in all of the blast results
for(file_name in SS_blast_files_gonads){
  
  #Read the file and store it in a new object
  file_data <- read.delim(file.path(blast_path, file_name), stringsAsFactors = FALSE, quote = "", sep = "\t", header = FALSE)

  #Add column names to the dataset
  colnames(file_data) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "reiro_prot_info", "prot_id", "reiro_prot_start", "reiro_prot_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Create a new object with a name based on the file name
  blast_name <- sub(".txt$", "", file_name) #Removes the file extension
  SS_blast_list_gonads[[blast_name]] <- file_data
}

```

Not that I have all of the BLAST files I am once again going to make sure they are filtered properly:

```{r filter-BLAST-dreiro-gonads}
#Use lapply to apply the function to each dataset stored in the list created above
blast_output_filtered_gonads <- lapply(SS_blast_list_gonads,
                                function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "reiro_prot_info", "prot_id", "reiro_prot_start", "reiro_prot_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID$trin_geneid){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- subset(data_frame, trin_geneid==gene)
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue == min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident == max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)

    # check that there is a single gene ID in scovelli, if not, only first row is kept
    if(length(gsub("^.*(GeneID:\\d+)\\].*$","\\1",uni_gene_subset_lowe_highpid$reiro_prot_info))>1){
    
    
      uni_gene_subset_lowe_highpid<-uni_gene_subset_lowe_highpid[1,]
    }
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
})

```

I then need to make sure the gene names and protein IDs are attached to each BLAST result:

```{r merge-genenames-gonads}

blast_output_merged_gonads <- lapply(blast_output_filtered_gonads,
                              function(dataframe){
  
   output <- merge(dataframe, unique(genes[,7:8]), by = "prot_id")
  
   return(output)
})

```

Lastly I need to save only the gene_id for all of the BLASTed genes to then upload to PANTHER for the gonads only.

```{r export-danio-geneID-gonads, eval=FALSE}

write.table(c(        
  blast_output_merged_gonads$SS_femGon_biased_TRgenes_reiro_blast$gene_name,
  blast_output_merged_gonads$SS_malGon_biased_TRgenes_reiro_blast$gene_name),
  'SS_GOnames_SBG_gonads.txt',
  sep = " ",
  quote=FALSE,
  row.names = FALSE,
  col.names = FALSE)

```

In PANTHER I uploaded the `SS_GOnames_SBG_gonads.txt` file, selected _Danio rerio_ as the organism, and then chose "Functional classification viewed in gene list" as the analysis type. 

I exported those results to a .txt file and will now be reading that into R and visualizing the results.

```{r panther-results-SBG-gonads}
#Read in the .txt file containing PANTHER results
panther <- read.delim("data/pantherGeneList_SBG_gonads.txt", 
                      header = FALSE)

#Add the approporiate column names for the PANTHER data
colnames(panther) <- c("GeneID", "MappedID", 
                       "GeneName", "pantherFAM", 
                       "panther_prot_class", "species")

#Add the identified PANTHER protein classes to the BLAST datasets
blast_output_merged_gonads$SS_femGon_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged_gonads$SS_femGon_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged_gonads$SS_malGon_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged_gonads$SS_malGon_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)

```

Now that we have the PANTHER protein classes attached to all of the successfully BLASTed sex-biased genes within each organ, we can plot the proportions of genes falling into each category to see what is most important for each of the organ types.

```{r plot-panther-SBG}
#Count the number of genes that are falling into each protein class
go_tables <- lapply(blast_output_merged_gonads, function(dat){
  
  #Create a table that counts the number of genes falling within each PANTHER prot. class
  panther_table <- data.frame(table(dat$panther_prot_class))
  
  #Make sure the protein names are in there as a character
  panther_table$Var1 <- as.character(panther_table$Var1)
  
  #Change all of the null results to say Unclassified
  panther_table$Var1[panther_table$Var1==""] <- "Unclassified"

  return(panther_table)
  })

#Merge all of the different organ data into one dataset
#Create an empty dataframe to store all of the values in
all_go_dat <- data.frame(matrix(ncol=4,nrow=0))

#Add column names to that empty dataset
colnames(all_go_dat) <- c(
  colnames(go_tables$SS_femGon_biased_TRgenes_reiro_blast),
  "bias_cat", "tissue")

#Create a vector of you tissues, in the same order that they appear in the go_table list
tissues <- c("Ovary", "Testes")

#Loop through all of the go_tables and combine into one
for(i in 1:length(go_tables)){
  #browser()
  tmp <- go_tables[[i]]
  if (nrow(tmp) == 0) {
    print(tmp)
  }else{
  
  tmp$bias_cat <- names(go_tables)[i]
  tmp$tissue <- tissues[[i]]
  all_go_dat <- rbind(all_go_dat, tmp)}
  
}

#Calculate the frequency at which each protein class occurs within the different tissue types
##Create an empty dataset to store the values in
all_go_sums <- data.frame(matrix(ncol = 4,
                                 nrow = 0))

##Add appropriate column names
colnames(all_go_sums) <- c("Freq", "prot_class", 
                           "tissue", "prop")

##Loop through each organ type to calculate frequencies
for(organ in unique(all_go_dat$tissue)){
  
  #subset the dataset based on the organ
  tmp <- all_go_dat[all_go_dat$tissue == organ, ]
  
  #Sum up all of the genes in each category
  go_sums <- as.data.frame(tapply(tmp$Freq, tmp$Var1, sum))
  
  #Add the protein class as a column rather than row names
  go_sums$prot_class <- rownames(go_sums)
  
  #Change the column name to "Frequency"
  colnames(go_sums)[1] <- "Freq"
  
  #Add the organ in a separate column and remove rownames
  go_sums$tissue <- organ
  rownames(go_sums) <- NULL
  
  #Calculate the respective proportion for each protein class
  go_sums$prop <- go_sums$Freq/sum(go_sums$Freq)
  
  #Export the data to the previously created dataframe
  all_go_sums <- rbind(all_go_sums, go_sums)
}

#Change all of the protein classes that have a frequency of less than 0.02 to "Other"
for(class in unique(all_go_sums$prot_class)){
  
  #Pull out the rows that contain that protein class
  match_rows <- which(all_go_sums$prot_class==class)
  
  #If any of the proportions associated with those rows is less than 0.02, classify as other
  if(any(all_go_sums$prop[match_rows]>0.01)==FALSE){
    
    all_go_sums$prot_class[match_rows] <- " other"
    
  }
}

#Change format of "Unclassified so it will come first"
all_go_sums$prot_class[all_go_sums$prot_class=="Unclassified"] <-  " unclassified"

#write.csv(all_go_sums, "data/GO_freq_SBG_Gonads.csv", row.names = FALSE)

#Generate the GO plot
organ_cols <- c("Ovary" = "#7fc97f", "Testes" = "#beaed4")

ggplot(all_go_sums, 
       aes(fct_rev(prot_class), prop, fill = tissue)) +   
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = organ_cols) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=20),
        axis.text.y = element_text(size=10),
        text=element_text(size=20),
        legend.position = "bottom") +
  coord_flip() + 
  labs(y="Proportion of sex-biased genes", x="")
```

For the over representation analysis I need the list of ovary and testis BLAST results in separate .txt files:

```{r export-danio-geneID-gonadsINDV, eval=FALSE}

write.table(blast_output_merged_gonads$SS_femGon_biased_TRgenes_reiro_blast$gene_name,
            'SS_GOnames_SBG_ovary.txt',
            sep = " ",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)

write.table(blast_output_merged_gonads$SS_malGon_biased_TRgenes_reiro_blast$gene_name,
            'SS_GOnames_SBG_testes.txt',
            sep = " ",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)
```